// @sap/cds-compiler/lib/gen/JSONParser.js
// Generated from JSON.g4 by ANTLR 4.6
// jshint ignore: start
var antlr4 = require('antlr4/index');


  function getLoc(ctx,ret) {
    if(ctx.start) {
      ret.start = {
        offset:ctx.start.start,
        start: ctx.start.start,
        stop: ctx.start.stop,
        line: ctx.start.line,
        column: ctx.start.column + 1
      }
    }
    if(ctx.stop) {
      ret.stop = {
        offset: ctx.stop.stop+1,
        start: ctx.stop.start,
        stop: ctx.stop.stop,
        line: ctx.stop.line,
        column: ctx.stop.column + 1 + (ctx.stop.stop - ctx.stop.start) // end of last token
      }
    }
  } // getLoc

  function unquote(s) {
    return JSON.parse(s); // unquote for free
  }


var grammarFileName = "JSON.g4";

var serializedATN = ["\u0003\u0430\ud6d1\u8206\uad2d\u4417\uaef1\u8d80\uaadd",
    "\u0003\u0010H\u0004\u0002\t\u0002\u0004\u0003\t\u0003\u0004\u0004\t",
    "\u0004\u0004\u0005\t\u0005\u0004\u0006\t\u0006\u0003\u0002\u0003\u0002",
    "\u0005\u0002\u000f\n\u0002\u0003\u0003\u0003\u0003\u0003\u0003\u0003",
    "\u0003\u0007\u0003\u0015\n\u0003\f\u0003\u000e\u0003\u0018\u000b\u0003",
    "\u0003\u0003\u0003\u0003\u0003\u0003\u0003\u0003\u0005\u0003\u001e\n",
    "\u0003\u0003\u0004\u0003\u0004\u0003\u0004\u0003\u0004\u0003\u0004\u0003",
    "\u0004\u0003\u0005\u0003\u0005\u0003\u0005\u0003\u0005\u0003\u0005\u0003",
    "\u0005\u0003\u0005\u0003\u0005\u0003\u0005\u0007\u0005/\n\u0005\f\u0005",
    "\u000e\u00052\u000b\u0005\u0003\u0005\u0003\u0005\u0003\u0005\u0003",
    "\u0005\u0005\u00058\n\u0005\u0003\u0006\u0003\u0006\u0003\u0006\u0003",
    "\u0006\u0003\u0006\u0003\u0006\u0003\u0006\u0003\u0006\u0003\u0006\u0003",
    "\u0006\u0003\u0006\u0003\u0006\u0005\u0006F\n\u0006\u0003\u0006\u0002",
    "\u0002\u0007\u0002\u0004\u0006\b\n\u0002\u0002M\u0002\u000e\u0003\u0002",
    "\u0002\u0002\u0004\u001d\u0003\u0002\u0002\u0002\u0006\u001f\u0003\u0002",
    "\u0002\u0002\b7\u0003\u0002\u0002\u0002\nE\u0003\u0002\u0002\u0002\f",
    "\u000f\u0005\u0004\u0003\u0002\r\u000f\u0005\b\u0005\u0002\u000e\f\u0003",
    "\u0002\u0002\u0002\u000e\r\u0003\u0002\u0002\u0002\u000f\u0003\u0003",
    "\u0002\u0002\u0002\u0010\u0011\u0007\u0003\u0002\u0002\u0011\u0016\u0005",
    "\u0006\u0004\u0002\u0012\u0013\u0007\u0004\u0002\u0002\u0013\u0015\u0005",
    "\u0006\u0004\u0002\u0014\u0012\u0003\u0002\u0002\u0002\u0015\u0018\u0003",
    "\u0002\u0002\u0002\u0016\u0014\u0003\u0002\u0002\u0002\u0016\u0017\u0003",
    "\u0002\u0002\u0002\u0017\u0019\u0003\u0002\u0002\u0002\u0018\u0016\u0003",
    "\u0002\u0002\u0002\u0019\u001a\u0007\u0005\u0002\u0002\u001a\u001e\u0003",
    "\u0002\u0002\u0002\u001b\u001c\u0007\u0003\u0002\u0002\u001c\u001e\u0007",
    "\u0005\u0002\u0002\u001d\u0010\u0003\u0002\u0002\u0002\u001d\u001b\u0003",
    "\u0002\u0002\u0002\u001e\u0005\u0003\u0002\u0002\u0002\u001f \u0007",
    "\f\u0002\u0002 !\b\u0004\u0001\u0002!\"\u0007\u0006\u0002\u0002\"#\u0005",
    "\n\u0006\u0002#$\b\u0004\u0001\u0002$\u0007\u0003\u0002\u0002\u0002",
    "%&\u0007\u0007\u0002\u0002&\'\b\u0005\u0001\u0002\'(\u0005\n\u0006\u0002",
    "(0\b\u0005\u0001\u0002)*\u0007\u0004\u0002\u0002*+\b\u0005\u0001\u0002",
    "+,\u0005\n\u0006\u0002,-\b\u0005\u0001\u0002-/\u0003\u0002\u0002\u0002",
    ".)\u0003\u0002\u0002\u0002/2\u0003\u0002\u0002\u00020.\u0003\u0002\u0002",
    "\u000201\u0003\u0002\u0002\u000213\u0003\u0002\u0002\u000220\u0003\u0002",
    "\u0002\u000234\u0007\b\u0002\u000248\u0003\u0002\u0002\u000256\u0007",
    "\u0007\u0002\u000268\u0007\b\u0002\u00027%\u0003\u0002\u0002\u00027",
    "5\u0003\u0002\u0002\u00028\t\u0003\u0002\u0002\u00029:\u0007\f\u0002",
    "\u0002:F\b\u0006\u0001\u0002;<\u0007\r\u0002\u0002<F\b\u0006\u0001\u0002",
    "=F\u0005\u0004\u0003\u0002>F\u0005\b\u0005\u0002?@\u0007\t\u0002\u0002",
    "@F\b\u0006\u0001\u0002AB\u0007\n\u0002\u0002BF\b\u0006\u0001\u0002C",
    "D\u0007\u000b\u0002\u0002DF\b\u0006\u0001\u0002E9\u0003\u0002\u0002",
    "\u0002E;\u0003\u0002\u0002\u0002E=\u0003\u0002\u0002\u0002E>\u0003\u0002",
    "\u0002\u0002E?\u0003\u0002\u0002\u0002EA\u0003\u0002\u0002\u0002EC\u0003",
    "\u0002\u0002\u0002F\u000b\u0003\u0002\u0002\u0002\b\u000e\u0016\u001d",
    "07E"].join("");


var atn = new antlr4.atn.ATNDeserializer().deserialize(serializedATN);

var decisionsToDFA = atn.decisionToState.map( function(ds, index) { return new antlr4.dfa.DFA(ds, index); });

var sharedContextCache = new antlr4.PredictionContextCache();

var literalNames = [ null, "'{'", "','", "'}'", "':'", "'['", "']'", "'true'", 
                     "'false'", "'null'" ];

var symbolicNames = [ null, null, null, null, null, null, null, null, null, 
                      null, "STRING", "NUMBER", "WHITESPACE", "MULTI_LINE_COMMENT", 
                      "SINGLE_LINE_COMMENT" ];

var ruleNames =  [ "json", "obj", "pair", "array", "value" ];

function JSONParser (input) {
	antlr4.Parser.call(this, input);
    this._interp = new antlr4.atn.ParserATNSimulator(this, atn, decisionsToDFA, sharedContextCache);
    this.ruleNames = ruleNames;
    this.literalNames = literalNames;
    this.symbolicNames = symbolicNames;
    return this;
}

JSONParser.prototype = Object.create(antlr4.Parser.prototype);
JSONParser.prototype.constructor = JSONParser;

Object.defineProperty(JSONParser.prototype, "atn", {
	get : function() {
		return atn;
	}
});

JSONParser.EOF = antlr4.Token.EOF;
JSONParser.T__0 = 1;
JSONParser.T__1 = 2;
JSONParser.T__2 = 3;
JSONParser.T__3 = 4;
JSONParser.T__4 = 5;
JSONParser.T__5 = 6;
JSONParser.T__6 = 7;
JSONParser.T__7 = 8;
JSONParser.T__8 = 9;
JSONParser.STRING = 10;
JSONParser.NUMBER = 11;
JSONParser.WHITESPACE = 12;
JSONParser.MULTI_LINE_COMMENT = 13;
JSONParser.SINGLE_LINE_COMMENT = 14;

JSONParser.RULE_json = 0;
JSONParser.RULE_obj = 1;
JSONParser.RULE_pair = 2;
JSONParser.RULE_array = 3;
JSONParser.RULE_value = 4;

function JsonContext(parser, parent, invokingState) {
	if(parent===undefined) {
	    parent = null;
	}
	if(invokingState===undefined || invokingState===null) {
		invokingState = -1;
	}
	antlr4.ParserRuleContext.call(this, parent, invokingState);
    this.parser = parser;
    this.ruleIndex = JSONParser.RULE_json;
    this.ret = {}
    return this;
}

JsonContext.prototype = Object.create(antlr4.ParserRuleContext.prototype);
JsonContext.prototype.constructor = JsonContext;

JsonContext.prototype.obj = function() {
    return this.getTypedRuleContext(ObjContext,0);
};

JsonContext.prototype.array = function() {
    return this.getTypedRuleContext(ArrayContext,0);
};




JSONParser.JsonContext = JsonContext;

JSONParser.prototype.json = function() {

    var localctx = new JsonContext(this, this._ctx, this.state);
    this.enterRule(localctx, 0, JSONParser.RULE_json);
    try {
        this.state = 12;
        this._errHandler.sync(this);
        switch(this._input.LA(1)) {
        case JSONParser.T__0:
            this.enterOuterAlt(localctx, 1);
            this.state = 10;
            this.obj(localctx.ret);
            break;
        case JSONParser.T__4:
            this.enterOuterAlt(localctx, 2);
            this.state = 11;
            this.array(localctx.ret);
            break;
        default:
            throw new antlr4.error.NoViableAltException(this);
        }
    } catch (re) {
    	if(re instanceof antlr4.error.RecognitionException) {
	        localctx.exception = re;
	        this._errHandler.reportError(this, re);
	        this._errHandler.recover(this, re);
	    } else {
	    	throw re;
	    }
    } finally {
        this.exitRule();
    }
    return localctx;
};

function ObjContext(parser, parent, invokingState, ret) {
	if(parent===undefined) {
	    parent = null;
	}
	if(invokingState===undefined || invokingState===null) {
		invokingState = -1;
	}
	antlr4.ParserRuleContext.call(this, parent, invokingState);
    this.parser = parser;
    this.ruleIndex = JSONParser.RULE_obj;
    this.ret = null
    this.ret = ret || null;
    return this;
}

ObjContext.prototype = Object.create(antlr4.ParserRuleContext.prototype);
ObjContext.prototype.constructor = ObjContext;

ObjContext.prototype.pair = function(i) {
    if(i===undefined) {
        i = null;
    }
    if(i===null) {
        return this.getTypedRuleContexts(PairContext);
    } else {
        return this.getTypedRuleContext(PairContext,i);
    }
};




JSONParser.ObjContext = ObjContext;

JSONParser.prototype.obj = function(ret) {

    var localctx = new ObjContext(this, this._ctx, this.state, ret);
    this.enterRule(localctx, 2, JSONParser.RULE_obj);
    localctx.ret.type="object"; localctx.ret.value=Object.create(null); getLoc(localctx,localctx.ret) 
    var _la = 0; // Token type
    try {
        this.state = 27;
        this._errHandler.sync(this);
        var la_ = this._interp.adaptivePredict(this._input,2,this._ctx);
        switch(la_) {
        case 1:
            this.enterOuterAlt(localctx, 1);
            this.state = 14;
            this.match(JSONParser.T__0);
            this.state = 15;
            this.pair(localctx.ret.value);
            this.state = 20;
            this._errHandler.sync(this);
            _la = this._input.LA(1);
            while(_la===JSONParser.T__1) {
                this.state = 16;
                this.match(JSONParser.T__1);
                this.state = 17;
                this.pair(localctx.ret.value);
                this.state = 22;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
            }
            this.state = 23;
            this.match(JSONParser.T__2);
            break;

        case 2:
            this.enterOuterAlt(localctx, 2);
            this.state = 25;
            this.match(JSONParser.T__0);
            this.state = 26;
            this.match(JSONParser.T__2);
            break;

        }
        this._ctx.stop = this._input.LT(-1);
        getLoc(localctx,localctx.ret)
    } catch (re) {
    	if(re instanceof antlr4.error.RecognitionException) {
	        localctx.exception = re;
	        this._errHandler.reportError(this, re);
	        this._errHandler.recover(this, re);
	    } else {
	    	throw re;
	    }
    } finally {
        this.exitRule();
    }
    return localctx;
};

function PairContext(parser, parent, invokingState, ret) {
	if(parent===undefined) {
	    parent = null;
	}
	if(invokingState===undefined || invokingState===null) {
		invokingState = -1;
	}
	antlr4.ParserRuleContext.call(this, parent, invokingState);
    this.parser = parser;
    this.ruleIndex = JSONParser.RULE_pair;
    this.ret = null
    this.val = {}
    this._STRING = null; // Token
    this.ret = ret || null;
    return this;
}

PairContext.prototype = Object.create(antlr4.ParserRuleContext.prototype);
PairContext.prototype.constructor = PairContext;

PairContext.prototype.STRING = function() {
    return this.getToken(JSONParser.STRING, 0);
};

PairContext.prototype.value = function() {
    return this.getTypedRuleContext(ValueContext,0);
};




JSONParser.PairContext = PairContext;

JSONParser.prototype.pair = function(ret) {

    var localctx = new PairContext(this, this._ctx, this.state, ret);
    this.enterRule(localctx, 4, JSONParser.RULE_pair);
    getLoc(localctx,localctx.val)
    try {
        this.enterOuterAlt(localctx, 1);
        this.state = 29;
        localctx._STRING = this.match(JSONParser.STRING);

        this.state = 31;
        this.match(JSONParser.T__3);
        this.state = 32;
        this.value(localctx.val);
        localctx.ret[unquote((localctx._STRING===null ? null : localctx._STRING.text))]=localctx.val
        this._ctx.stop = this._input.LT(-1);
        getLoc(localctx,localctx.val)
    } catch (re) {
    	if(re instanceof antlr4.error.RecognitionException) {
	        localctx.exception = re;
	        this._errHandler.reportError(this, re);
	        this._errHandler.recover(this, re);
	    } else {
	    	throw re;
	    }
    } finally {
        this.exitRule();
    }
    return localctx;
};

function ArrayContext(parser, parent, invokingState, ret) {
	if(parent===undefined) {
	    parent = null;
	}
	if(invokingState===undefined || invokingState===null) {
		invokingState = -1;
	}
	antlr4.ParserRuleContext.call(this, parent, invokingState);
    this.parser = parser;
    this.ruleIndex = JSONParser.RULE_array;
    this.ret = null
    this.val = null
    this.ret = ret || null;
    return this;
}

ArrayContext.prototype = Object.create(antlr4.ParserRuleContext.prototype);
ArrayContext.prototype.constructor = ArrayContext;

ArrayContext.prototype.value = function(i) {
    if(i===undefined) {
        i = null;
    }
    if(i===null) {
        return this.getTypedRuleContexts(ValueContext);
    } else {
        return this.getTypedRuleContext(ValueContext,i);
    }
};




JSONParser.ArrayContext = ArrayContext;

JSONParser.prototype.array = function(ret) {

    var localctx = new ArrayContext(this, this._ctx, this.state, ret);
    this.enterRule(localctx, 6, JSONParser.RULE_array);
    localctx.ret.type="array"; localctx.ret.value={}; getLoc(localctx,localctx.ret); let ID=0;
    var _la = 0; // Token type
    try {
        this.state = 53;
        this._errHandler.sync(this);
        var la_ = this._interp.adaptivePredict(this._input,4,this._ctx);
        switch(la_) {
        case 1:
            this.enterOuterAlt(localctx, 1);
            this.state = 35;
            this.match(JSONParser.T__4);
            localctx.val={}
            this.state = 37;
            this.value(localctx.val);
            localctx.ret.value[ID++]=localctx.val
            this.state = 46;
            this._errHandler.sync(this);
            _la = this._input.LA(1);
            while(_la===JSONParser.T__1) {
                this.state = 39;
                this.match(JSONParser.T__1);
                localctx.val={}
                this.state = 41;
                this.value(localctx.val);
                localctx.ret.value[ID++]=localctx.val
                this.state = 48;
                this._errHandler.sync(this);
                _la = this._input.LA(1);
            }
            this.state = 49;
            this.match(JSONParser.T__5);
            break;

        case 2:
            this.enterOuterAlt(localctx, 2);
            this.state = 51;
            this.match(JSONParser.T__4);
            this.state = 52;
            this.match(JSONParser.T__5);
            break;

        }
        this._ctx.stop = this._input.LT(-1);
        getLoc(localctx,localctx.ret)
    } catch (re) {
    	if(re instanceof antlr4.error.RecognitionException) {
	        localctx.exception = re;
	        this._errHandler.reportError(this, re);
	        this._errHandler.recover(this, re);
	    } else {
	    	throw re;
	    }
    } finally {
        this.exitRule();
    }
    return localctx;
};

function ValueContext(parser, parent, invokingState, ret) {
	if(parent===undefined) {
	    parent = null;
	}
	if(invokingState===undefined || invokingState===null) {
		invokingState = -1;
	}
	antlr4.ParserRuleContext.call(this, parent, invokingState);
    this.parser = parser;
    this.ruleIndex = JSONParser.RULE_value;
    this.ret = null
    this._STRING = null; // Token
    this._NUMBER = null; // Token
    this.ret = ret || null;
    return this;
}

ValueContext.prototype = Object.create(antlr4.ParserRuleContext.prototype);
ValueContext.prototype.constructor = ValueContext;

ValueContext.prototype.STRING = function() {
    return this.getToken(JSONParser.STRING, 0);
};

ValueContext.prototype.NUMBER = function() {
    return this.getToken(JSONParser.NUMBER, 0);
};

ValueContext.prototype.obj = function() {
    return this.getTypedRuleContext(ObjContext,0);
};

ValueContext.prototype.array = function() {
    return this.getTypedRuleContext(ArrayContext,0);
};




JSONParser.ValueContext = ValueContext;

JSONParser.prototype.value = function(ret) {

    var localctx = new ValueContext(this, this._ctx, this.state, ret);
    this.enterRule(localctx, 8, JSONParser.RULE_value);
    getLoc(localctx,localctx.ret)
    try {
        this.state = 67;
        this._errHandler.sync(this);
        switch(this._input.LA(1)) {
        case JSONParser.STRING:
            this.enterOuterAlt(localctx, 1);
            this.state = 55;
            localctx._STRING = this.match(JSONParser.STRING);
            localctx.ret.type="string"; localctx.ret.value=unquote((localctx._STRING===null ? null : localctx._STRING.text));
            break;
        case JSONParser.NUMBER:
            this.enterOuterAlt(localctx, 2);
            this.state = 57;
            localctx._NUMBER = this.match(JSONParser.NUMBER);
            localctx.ret.type="number"; localctx.ret.value=unquote((localctx._NUMBER===null ? null : localctx._NUMBER.text));
            break;
        case JSONParser.T__0:
            this.enterOuterAlt(localctx, 3);
            this.state = 59;
            this.obj(localctx.ret);
            break;
        case JSONParser.T__4:
            this.enterOuterAlt(localctx, 4);
            this.state = 60;
            this.array(localctx.ret);
            break;
        case JSONParser.T__6:
            this.enterOuterAlt(localctx, 5);
            this.state = 61;
            this.match(JSONParser.T__6);
            localctx.ret.type="boolean"; localctx.ret.value=true;
            break;
        case JSONParser.T__7:
            this.enterOuterAlt(localctx, 6);
            this.state = 63;
            this.match(JSONParser.T__7);
            localctx.ret.type="boolean"; localctx.ret.value=false;
            break;
        case JSONParser.T__8:
            this.enterOuterAlt(localctx, 7);
            this.state = 65;
            this.match(JSONParser.T__8);
            localctx.ret.type="null"; localctx.ret.value=null;
            break;
        default:
            throw new antlr4.error.NoViableAltException(this);
        }
        this._ctx.stop = this._input.LT(-1);
        getLoc(localctx,localctx.ret)
    } catch (re) {
    	if(re instanceof antlr4.error.RecognitionException) {
	        localctx.exception = re;
	        this._errHandler.reportError(this, re);
	        this._errHandler.recover(this, re);
	    } else {
	    	throw re;
	    }
    } finally {
        this.exitRule();
    }
    return localctx;
};


exports.JSONParser = JSONParser;
