// @sap/odata-v4/lib/errors/ApplicationError.js
'use strict';

/**
 * Error that can be thrown by an application using the OData library.
 * It provides methods to set additional properties as defined in the OData format specifications
 * that are used elsewhere, e.g., by the error serializers, and are therefore not to be renamed.
 * @extends Error
 */
class ApplicationError extends Error {

    /**
     * @param {string} message
     */
    constructor(message) {
        if (!message || typeof message !== 'string') throw new Error("Parameter 'message' must be a string");
        super(message);

        // Overwrite the JavaScript standard Error property 'name'.
        this.name = 'ApplicationError';
    }

    /**
     * Sets the root error if there is any.
     * @param {?Error} rootCause the root cause
     * @returns {ApplicationError} this instance
     */
    setRootCause(rootCause) {
        if (rootCause && !(rootCause instanceof Error)) {
            throw new Error("Parameter 'rootCause' must be an instance of Error");
        }
        this._rootCause = rootCause;
        return this;
    }

    /**
     * @returns {?Error} the root cause
     */
    getRootCause() {
        return this._rootCause;
    }

    /**
     * Set the HTTP status code (default is 500 - Internal Server Error).
     * @param {?(HttpStatusCode.StatusCodes|number)} statusCode the status code
     * @returns {ApplicationError} this instance
     */
    setStatusCode(statusCode) {
        if (statusCode && typeof statusCode !== 'number') throw new Error("Parameter 'statusCode' must be a number");
        this.statusCode = statusCode;
        return this;
    }

    /**
     * Set the service-defined error code. This code serves as a sub-status for the HTTP error code.
     * @param {?string} code the error code
     * @returns {ApplicationError} this instance
     */
    setCode(code) {
        if (code && typeof code !== 'string') throw new Error("Parameter 'code' must be a string");
        this.code = code;
        return this;
    }

    /**
     * Set the target of this error (for example, the name of the property in error).
     * @param {?string} target the target
     * @returns {ApplicationError} this instance
     */
    setTarget(target) {
        if (target && typeof target !== 'string') throw new Error("Parameter 'target' must be a string");
        this.target = target;
        return this;
    }

    /**
     * Set additional details.
     * @param {?Array<{ message: string, code: ?string, target: ?string }>} details the details
     * @returns {ApplicationError} this instance
     */
    setDetails(details) {
        if (details) {
            if (!Array.isArray(details)) throw new Error("Parameter 'details' must be an Array");
            for (const d of details) {
                if (typeof d !== 'object') throw new Error('Detail must be an Object');
                if (!d.message || typeof d.message !== 'string') throw new Error('Detail message must be a string');
                if (d.code && typeof d.code !== 'string') throw new Error('Detail code must be a string');
                if (d.target && typeof d.target !== 'string') throw new Error('Detail target must be a string');
            }
        }
        this.details = details;
        return this;
    }

    /**
     * Add an annotation to this error (for example, a severity classification).
     * @param {string} annotation the annotation in the form '@<namespace>.<term>'
     * @param {*} value the value of the annotation
     * @returns {ApplicationError} this instance
     */
    addAnnotation(annotation, value) {
        if (typeof annotation !== 'string') throw new Error("Parameter 'annotation' must be a string");
        if (!annotation.startsWith('@')) throw new Error("Parameter 'annotation' must start with '@'");
        if (!annotation.includes('.')) throw new Error("Parameter 'annotation' must contain a '.'");
        this[annotation] = value;
        return this;
    }
}

module.exports = ApplicationError;
