// @sap/odata-v4/lib/errors/UriQueryOptionSemanticError.js
'use strict';

const util = require('util');
const UriSemanticError = require('./UriSemanticError');

/**
 * UriQueryOptionSemanticError which is mainly thrown by the UriParser when there is a problem with
 * interpreting a query option of the incoming url.
 *
 * @extends UriSemanticError
 */
class UriQueryOptionSemanticError extends UriSemanticError {

    /**
     * Creating an instance of UriQueryOptionSemanticError.
     *
     * @param {string} message The message of the error
     * @param {...string} parameters parameters for the message
     */
    constructor(message, ...parameters) {
        super(util.format(message, ...parameters));
        this.name = UriSemanticError.ErrorNames.URI_QUERY_OPTION_SEMANTIC;
    }
}

UriQueryOptionSemanticError.Message = {
    ENTITY_SET_OR_SINGLETON_NOT_FOUND: "Expected '%s' to be an entity set or a singleton",
    NAVIGATION_PROPERTY_NOT_FOUND: "Navigation property '%s' is not defined in type '%s'",
    PROPERTY_MUST_FOLLOW_STRUCTURED_TYPE: "Property '%s' must follow a structured type",
    COMPLEX_TYPE_NOT_FOUND: "Complex type '%s' not found",
    ENUM_TYPE_NOT_FOUND: "Enumeration type '%s' not found",
    REFERENCED_NOT_TYPED: 'The referenced part is not typed',
    WRONG_NAMESPACE: "Wrong namespace '%s'",
    TYPE_LITERAL: "Type literal expected in '%s' at position %d",
    TYPE_CAST_NOT_CHAINABLE: "The type cast '%s' is not chainable after '%s'",
    KEY_NOT_ALLOWED: 'A key is not allowed on non-collection navigation properties',
    DUPLICATE_PARAMETER: "Duplicate function parameter '%s'",
    PARAMETER_NULL: "Function parameter '%s' must not be null",
    NOT_COMPOSABLE: "Function '%s' is not composable",
    COLLECTION: 'Collection not allowed',
    OPTION_NOT_ALLOWED: "Query option '%s' is not supported for this request",

    IS_PROPERTY: "Alias '%s' is already a property.",
    CUSTOM_AGGREGATION_METHOD_NOT_FOUND: "Custom aggregation method '%s' is not defined",
    ONLY_FOR_PRIMITIVE_TYPES: "'%s' expressions must return primitive values",
    FUNCTION_MUST_USE_COLLECTIONS:
        "Function '%s' must have entity- or complex-collection binding parameter and return type"
};

module.exports = UriQueryOptionSemanticError;
