// @sap/odata-v4/lib/errors/UriSemanticError.js
'use strict';

const util = require('util');
const AbstractError = require('./AbstractError');

/**
 * UriSemanticError which is mainly thrown by the UriParser when there is a problem with
 * interpreting the incoming url.
 *
 * @extends AbstractError
 */
class UriSemanticError extends AbstractError {

    /**
     * Creating an instance of UriSemanticError.
     *
     * @param {string} message The message of the error
     * @param {...string} parameters parameters for the message
     */
    constructor(message, ...parameters) {
        super(AbstractError.ErrorNames.URI_SEMANTIC, util.format(message, ...parameters));
    }
}

UriSemanticError.Message = {
    WRONG_NAME: "'%s' is not an entity set, a singleton, an action import, or a function import",
    PROPERTY_NOT_FOUND: "Property '%s' does not exist in type '%s'",
    ENTITY_SET_NOT_FOUND: "Expected '%s' to be an entity set",
    ENTITY_TYPE_NOT_FOUND: "Entity type '%s' not found",
    NEITHER_STRUCTURED_TYPE_NOR_BOUND_OPERATION: "'%s' is neither a structured type nor a bound operation",
    FUNCTION_NOT_FOUND: "No function for '%s' found for parameter names '%s'",
    KEY_PREDICATES_MISSING: "Missing key predicates '%s'",
    DUPLICATE_KEY: "Duplicate key '%s' is not allowed",
    WRONG_KEY_NAME: "'%s' is not a valid key",
    WRONG_VALUE: "The type '%s' does not allow '%s' as its value",
    INCOMPATIBLE_TYPE: "The type '%s' is not compatible to '%s'"
};

module.exports = UriSemanticError;
