// @sap/cds-services/lib/adapter/odata-v4/utils/logger.js
const logLevels = {
  error: 0,
  warn: 1,
  info: 2,
  verbose: 3,
  debug: 4,
  silly: 5
}

const _nameToLevel = (name) => {
  if (logLevels[name]) {
    return logLevels[name]
  }

  return (name === 'path') ? 3 : 0
}

const _nameToLoggerName = (name) => {
  switch (name) {
    case 'fatal':
    case 'warn':
      return 'error'
    case 'debug':
    case 'silly':
      return 'info'
    case 'verbose':
    case 'path':
      return 'dir'
    default:
      return name
  }
}

const _shouldLog = (level, name) => {
  return (logLevels.hasOwnProperty(level) && logLevels[level] >= _nameToLevel(name))
}

const _createLogger = (level, name) => {
  if (_shouldLog(level, name)) {
    const loggerName = _nameToLoggerName(name)

    if (console[loggerName]) {
      return (logEntry) => {
        console[loggerName](`${name}: ` + logEntry)
      }
    }
  }

  return () => {}
}

const getDefaultLogger = (level) => {
  const logger = {}
  logger.debug = _createLogger(level, 'debug')
  logger.path = _createLogger(level, 'path')
  logger.info = _createLogger(level, 'info')
  logger.warn = _createLogger(level, 'warn')
  logger.warning = logger.warn
  logger.error = _createLogger(level, 'error')
  logger.fatal = _createLogger(level, 'fatal')

  return logger
}

const _defaultsForLogger = ({logger, logLevel}) => {
  const defaultLogger = getDefaultLogger(logLevel)

  for (const key of Object.keys(defaultLogger)) {
    if (typeof logger[key] === 'function' && _shouldLog(logLevel, key)) {
      defaultLogger[key] = logger[key]
    }
  }

  return defaultLogger
}

/**
 * Get a logger to be used in the odata library.
 *
 * @param {Object} [options] - optional object with options.
 * @param {Object} [options.logger] - optional custom logger.
 * @param {string} [options.logLevel] - optional log level to be used according to winston/npm specification.
 * @returns {Object} - Logger to be used in the odata library.
 * @private
 */
const getLogger = (options) => {
  if (!options) {
    options = {}
  }

  options.logLevel = options.logLevel || 'warn'

  // if a custom logger is provided, then use it.
  if (options.logger) {
    return _defaultsForLogger(options)
  }

  return getDefaultLogger(options.logLevel)
}

module.exports = getLogger
