// @sap/odata-v4/lib/core/PlainHttpResponse.js
'use strict';

const CRLF = '\r\n';
const HTTP_VERSION = 'HTTP/1.1';

const HttpStatusCode = require('../http/HttpStatusCode');
const HttpStatusCodes = HttpStatusCode.StatusCodes;
const HttpStatusCodesText = HttpStatusCode.Texts;
const Writable = require('stream').Writable;

/**
 * Simulates an http.ServerResponse
 * Used when parsing an multipart/mixed request into sub requests
 *
 * @extends Writable
 * @hideconstructor
 */
class PlainHttpResponse extends Writable {

    /**
     * Constructor
     */
    constructor() {
        super();
        this.statusCode = HttpStatusCodes.OK;
        this._buffers = [];
        this._header = {};

        // as in nodejs
        this.headersSent = false;
        this._odataRequestId = null;
    }


    /**
     * Return  OData request id
     *
     * @returns {?string}
     */
    getOdataRequestId() {
        return this._odataRequestId;
    }

    /**
     * Set OData request id
     *
     * @param {string} id OData request id
     * @returns {OdataRequest}
     */
    setOdataRequestId(id) {
        this._odataRequestId = id;
        return this;

    }

    /**
     * Writable implementation
     *
     * @param chunk
     * @param encoding
     * @param callback
     * @private
     */
    _write(chunk, encoding, callback) {
        this._buffers.push(chunk);

        callback();
    }


    /**
     * Returns the value of the requested header
     *
     * @param {string} name Header name
     * @returns {string} Header value
     */
    getHeader(name) {
        return this._header[name.toLowerCase()];
    }

    /**
     * Returns headers object
     *
     * @returns {Object}
     */
    getHeaders() {
        return this._header;
    }


    /**
     * Sets a header
     * @param {string} name
     * @param {string} value
     * @package
     */
    setHeader(name, value) {
        this._header[name.toLowerCase()] = value;
    }

    /**
     * Returns the http version
     * @returns {string}
     */
    getHttpVersion() {
        return HTTP_VERSION;
    }

    /**
     * Writes the response into an stream
     *
     * @param {Stream} stream
     * @package
     */
    writeTo(stream) {
        // write status line
        const statusCode = this.statusCode || HttpStatusCodes.OK;
        const statusCodeText = HttpStatusCodesText[statusCode];

        stream.write(this.getHttpVersion() + ' ' + statusCode + ' ' + statusCodeText + CRLF);

        // write headers
        for (let name of Object.keys(this._header)) {
            if (name.toLowerCase() !== 'content-length') {
                stream.write(name + ': ' + this._header[name] + CRLF);
            }
        }

        // stream.write('content-length:  ' + l.toString() + CRLF); // TODO check if content-length is a must
        stream.write(CRLF); // header body separator

        // write payload
        for (const buf of this._buffers) {
            stream.write(buf);
        }
    }
}

module.exports = PlainHttpResponse;
