// @sap/odata-v4/lib/serializer/ErrorJsonSerializer.js
'use strict';

const ErrorSerializer = require('./ErrorSerializer');

/**
 * The ErrorJsonSerializer serializes an error to an OData error in JSON format.
 *
 * @extends ErrorSerializer
 */
class ErrorJsonSerializer extends ErrorSerializer {

    /**
     * Creates an instance of ErrorJsonSerializer.
     *
     * @param {Error} error The error to serialize.
     * @throws {IllegalArgumentError} Thrown if provided error is undefined or does not have the correct structure
     */
    constructor(error) {
        super(error);
    }

    /**
     * Serializes the provided error to an OData error in JSON format.
     *
     * @returns {string} an OData error JSON string
     */
    serialize() {
        function addAnnotations(source, target) {
            let t = target;
            for (const name of Object.keys(source).filter(n => n.startsWith('@'))) t[name] = source[name];
        }

        let result = {
            code: this._error.code || null,
            message: this._error.message,
            target: this._error.target
        };

        if (this._error.details) {
            result.details = this._error.details.map(d => {
                let detail = {
                    code: d.code,
                    message: d.message,
                    target: d.target
                };
                addAnnotations(d, detail);
                return detail;
            });
        }

        addAnnotations(this._error, result);

        return JSON.stringify({ error: result });
    }
}

module.exports = ErrorJsonSerializer;
