// @sap/odata-v4/lib/serializer/MultipartWriter.js
'use strict';

const CRLF = '\r\n';

const ContentTypes = require('../format/ContentTypeInfo').ContentTypes;
const HeaderNames = require('../http/HttpHeader').HeaderNames;
const SerializationError = require('../errors/SerializationError');

const STATES = {
    WRITING_PAYLOAD_HEADER: 0,
    WRITING_OBJECT_HEADER: 1,
    WRITING_FIRST_PART: 2,
    WRITING_ENCAPSULATION: 3,
    WRITING_EPILOGUE: 4
};

/**
 * Writes content in application/multipart format into an stream.
 * Used to write OData batch responses. For nested multipart (e.g. for change set) one stream can be used with 2 MultipartWriter instances
 *
 * Sample call sequence:
 * - const w = new MultipartWriter(httpResponse)
 * - w->writePreamble (...)
 * - w->writePart (...)
 * - w->writePart (...)
 * - w->writeEpilogue (...)
 * - w->writeEnd
 *
 *
 * 'Content-Type: multipart/mixed; boundary=changeset', <-- written Automatically
 * 'CustomHeader: Value                                 <-- written by MultipartWriter->writeHeader
 * '',
 * '--changeset',                                       <-- written by MultipartWriter->writePart
 * 'Content-Type: application/http',
 * 'Content-ID: 1',                                     <-- extracted from the response headers
 * '',
 * 'POST /service/Customers HTTP/1.1',
 * 'Host: host',
 * 'Content-Type: application/json',
 * '',
 * '{}',
 * '--changeset',  <-- written by MultipartWriter->writePart
 * 'Content-Type: application/http',
 * 'Content-ID: 2',
 * '',
 * "PATCH /service/Customers('ALFKI') HTTP/1.1'",
 * 'Host: host',
 * 'Content-Type: application/json',
 * '',
 * '{}',
 * '--changeset--',  <-- written by MultipartWriter->writeEnd
 */
class MultipartWriter {

    /**
     * Constructor
     *
     * @param {Stream} writableStream Destination stream
     * @param {boolean} writeHeaderAsPayload If false then the headers are set into the destination stream via setHeader. For example if the
     * destination stream is a http.ServerResponse. If true the headers are serialised as name value pair into the payload.
     * @param {string} boundary Boundary for the multipart
     */
    constructor(writableStream, writeHeaderAsPayload, boundary) {

        this._state = writeHeaderAsPayload ? STATES.WRITING_PAYLOAD_HEADER : STATES.WRITING_OBJECT_HEADER;

        this._writableStream = writableStream;

        this._boundary = boundary;

    }

    /**
     * Writes a http header (name, value) into the destination stream
     *
     * @param {string} name Header name
     * @param {string} value Header value
     */
    writeHeader(name, value) {
        if (this._state === STATES.WRITING_PAYLOAD_HEADER) {
            this._writableStream.write(name + ': ' + value + CRLF);
        } else if (this._state === STATES.WRITING_OBJECT_HEADER) {
            this._writableStream.setHeader(name, value);
        } else {
            throw new SerializationError('MultipartWriter: header must be written first');
        }
    }

    /**
     * Writes the content type header (name, value) into the destination stream
     */
    writeContentType() {
        if (this._state === STATES.WRITING_PAYLOAD_HEADER) {
            this._writableStream.write(HeaderNames.CONTENT_TYPE + ': ' + ContentTypes.MULTIPART_MIXED +
                '; boundary=' + this._boundary + CRLF, 'utf8');
        } else {
            this._writableStream.setHeader(HeaderNames.CONTENT_TYPE, ContentTypes.MULTIPART_MIXED +
                '; boundary=' + this._boundary);
        }
    }


    /**
     * Writes the preamble into the stream
     *
     * @param {Buffer} buffer Preamble
     */
    writePreamble(buffer) {
        if (this._state === STATES.WRITING_PAYLOAD_HEADER ||
            this._state === STATES.WRITING_OBJECT_HEADER) {

            this.writeContentType();

            // write header separator

            if (this._state === STATES.WRITING_PAYLOAD_HEADER) {
                this._writableStream.write(CRLF, 'utf8');
            }
            this._state = STATES.PREAMBLE;

        }

        this._writableStream.write(buffer);
    }

    /**
     * Start writing the part (especially LF and boundaries)
     *
     * @throws Error
     */
    startPart() {
        if (this._state === STATES.WRITING_PAYLOAD_HEADER ||
            this._state === STATES.WRITING_OBJECT_HEADER) {

            this.writeContentType();

            if (this._state === STATES.WRITING_PAYLOAD_HEADER) {
                this._writableStream.write(CRLF, 'utf8');
            }
            this._state = STATES.WRITING_FIRST_PART;

            this.writeBoundary();
        } else if (
            this._state === STATES.PREAMBLE) {

            this._writableStream.write(CRLF, 'utf8');

            this.writeBoundary();
        } else if (
            this._state === STATES.WRITING_ENCAPSULATION) {

            this._writableStream.write(CRLF, 'utf8');

            this.writeBoundary();
        } else {
            throw new SerializationError('MultipartWriter: part could not be written');
        }
    }

    /**
     * Set state to writing encapsulation indicating that the closing boundary has been written
     */
    endPart() {
        this._state = STATES.WRITING_ENCAPSULATION;
    }

    /**
     * Write a buffer in a part
     * @param {Buffer} buffer Buffer with data to be written
     */
    writeIntoPart(buffer) {
        this._writableStream.write(buffer);
    }


    /**
     * Writes the closing bounder into the stream
     *
     * @param {boolean} last If true and additional CRLF is added after the end boundary (required for backward compatibility)
     */
    writeEndBoundary(last) {
        this._writableStream.write(CRLF + '--' + this._boundary + '--' + (last ? CRLF : ''));
    }


    /**
     * Writes the epilogue into the stream
     *
     * @param {Buffer} buffer Preamble
     */
    writeEpilogue(buffer) {
        if (this._state !== STATES.WRITING_ENCAPSULATION) {
            throw new SerializationError('MultipartWriter: part required before writing epilogue');
        }
        this.writeEndBoundary();

        // this.startData();
        this._writableStream.write(CRLF, 'utf8');
        this._writableStream.write(buffer);
        this._state = STATES.WRITING_EPILOGUE;
    }

    /**
     * Writes a boundary delimiter into the stream
     */
    writeBoundary() {
        const out = this._writableStream;
        out.write('--' + this._boundary + CRLF);
    }
}

module.exports = MultipartWriter;
