// @sap/odata-v4/index.js
'use strict';
/* eslint-disable global-require */

module.exports = {
    FullQualifiedName: require('./lib/FullQualifiedName'),
    ServiceFactory: require('./lib/ServiceFactory'),
    edm: require('./lib/edm'),
    uri: {
        UriResource: require('./lib/uri/UriResource'),
        Expression: require('./lib/uri/Expression'),
        BinaryExpression: require('./lib/uri/BinaryExpression'),
        UnaryExpression: require('./lib/uri/UnaryExpression'),
        MethodExpression: require('./lib/uri/MethodExpression'),
        Transformation: require('./lib/uri/apply/Transformation'),
        BottomTopTransformation: require('./lib/uri/apply/BottomTopTransformation'),
        AggregateExpression: require('./lib/uri/apply/AggregateExpression')
    },
    ApplicationError: require('./lib/errors/ApplicationError'),

    QueryOptions: require('./lib/uri/UriInfo').QueryOptions,
    HttpMethods: require('./lib/http/HttpMethod').Methods,
    HttpStatusCodes: require('./lib/http/HttpStatusCode').StatusCodes,
    PreferenceReturnValues: require('./lib/http/Preferences').ReturnValues,
    Components: require('./lib/core/ComponentManager').Components,
    ContentTypes: require('./lib/format/ContentTypeInfo').ContentTypes,
    BatchExitHandler: require('./lib/batch/BatchExitHandler')
};
