// @sap/odata-v4/lib/csdl/CsdlProperty.js
'use strict';

const validateThat = require('../validator/ParameterValidator').validateThat;
const IllegalArgumentError = require('../errors/IllegalArgumentError');

/**
 * <a href="./../ODataSpecification/odata-v4.0-errata03-os/complete/part3-csdl/odata-v4.0-errata03-os-part3-csdl-complete.html#_Toc453752525">
 *     OData CSDL # 6.1 Element edm:Property
 * </a>
 */
class CsdlProperty {

    /**
     * @param {string} name - OData CSDL # 6.1.1 Attribute Name
     * @param {FullQualifiedName} type - OData CSDL # 6.1.2 Attribute Type
     */
    constructor(name, type) {
        validateThat('name', name).truthy().typeOf('string');
        validateThat('type', type).truthy().instanceOf(Object);

        /**
         * OData CSDL # 6.1.1 Attribute Name
         * @type {string}
         */
        this.name = name;

        /**
         * OData CSDL # 6.1.2 Attribute Type
         * @type {FullQualifiedName}
         */
        this.type = type;

        /**
         * Indicates whether the property represents a collection
         * @type {boolean}
         */
        this.isCollection = false;

        // Facets

        /**
         * OData CSDL # 6.2.1 Attribute Nullable
         * @type {boolean}
         */
        this.isNullable = true;

        /**
         * OData CSDL # 6.2.2 Attribute MaxLength
         * @type {number | string}
         */
        this.maxLength = null;

        /**
         * OData CSDL # 6.2.3 Attribute Precision
         * @type {number}
         */
        this.precision = null;

        /**
         * OData CSDL # 6.2.4 Attribute Scale
         * @type {number | string}
         */
        this.scale = null;

        /**
         * OData CSDL # 6.2.7 Attribute DefaultValue
         */
        this.defaultValue = null;

        /**
         * OData CSDL # # 14.3 Element edm:Annotation
         * @type {CsdlAnnotation[]}
         */
        this.annotations = [];
    }

    /**
     * Sets whether the property represents a collection
     *
     * @param {boolean} isCollection - indicates, whether the property should represent a collection
     * @returns {CsdlProperty} this instance
     */
    setCollection(isCollection) {
        validateThat('isCollection', isCollection).notNullNorUndefined().typeOf('boolean');

        this.isCollection = isCollection;
        return this;
    }

    /**
     * Sets 'Nullable' facet.
     * OData CSDL # 6.2.1 Attribute Nullable
     *
     * @param {boolean} isNullable - value of the 'Nullable' facet
     * @returns {CsdlProperty} this instance
     */
    setNullable(isNullable) {
        validateThat('isNullable', isNullable).notNullNorUndefined().typeOf('boolean');

        this.isNullable = isNullable;
        return this;
    }

    /**
     * Sets 'MaxLength' facet.
     * OData CSDL # 6.2.2 Attribute MaxLength
     *
     * @param {number | string} maxLength - value of the 'MaxLength' facet. It must be either a
     * positive integer value or a special 'max' string value. More information can be found in the
     * CSDL specification.
     * @returns {CsdlProperty} this instance
     */
    setMaxLength(maxLength) {
        const validate = validateThat('maxLength', maxLength).notNullNorUndefined();

        // Check if the special 'max' value is specified.
        if (typeof maxLength === 'string') {
            // Check whether a string value is specified but not the allowed one.
            if (maxLength !== 'max') {
                throw IllegalArgumentError.createForIllegalTypeValue('maxLength', 'string', 'max');
            }
        } else {
            // Otherwise the value must be a positive integer.
            validate.integer().positiveInteger();
        }

        this.maxLength = maxLength;
        return this;
    }

    /**
     * Sets Precision facet.
     * OData CSDL # 6.2.3 Attribute Precision
     *
     * @param {number} precision - value of the Precision facet.
     * @returns {CsdlProperty} this instance
     */
    setPrecision(precision) {
        validateThat('precision', precision)
            .notNullNorUndefined()
            .integer()
            .nonNegativeInteger();

        this.precision = precision;
        return this;
    }

    /**
     * Sets 'Scale' facet.
     * OData CSDL # 6.2.4 Attribute Scale
     *
     * @param {number | string} scale - value of the 'Scale' facet. It must be either a
     * non-negative integer value or a special 'variable' string value. More information can be
     * found in the CSDL specification.
     * @returns {CsdlProperty} this instance
     */
    setScale(scale) {
        const validate = validateThat('scale', scale).notNullNorUndefined();

        // Check if the special 'variable' value is specified.
        if (typeof scale === 'string') {
            // Check whether a string value is specified but not the allowed one.
            if (scale !== 'variable') {
                throw IllegalArgumentError.createForIllegalTypeValue('scale', 'string', 'variable');
            }
        } else {
            // Otherwise the value must be a non-negative integer.
            validate.integer().nonNegativeInteger();
        }

        this.scale = scale;
        return this;
    }

    /**
     * OData CSDL # 6.2.5 Attribute Unicode.
     * The Unicode attribute for properties is not supported
     *
     * The property always has value 'true'. Modification of the property value is not supported.
     * @type {boolean}
     */
    get unicode() {
        return true;
    }

    /**
     * OData CSDL # 6.2.6 Attribute SRID.
     * The property always has an undefined value. Modification of the property value is not supported.
     * @type {number | string}
     */
    get srid() {
        return undefined;
    }

    /**
     * Sets default value.
     * OData CSDL # 6.2.7 Attribute DefaultValue
     *
     * @param {*} defaultValue - value, which the property should have by default
     * @returns {CsdlProperty} this instance
     */
    setDefaultValue(defaultValue) {
        this.defaultValue = defaultValue;
        return this;
    }

    /**
     * Sets a list of annotations for the property.
     * OData CSDL # 14.3 Element edm:Annotation
     *
     * @param {CsdlAnnotation[]} annotations - annotations for the property
     * @returns {CsdlProperty} this instance
     */
    setAnnotations(annotations) {
        validateThat('annotations', annotations)
            .notNullNorUndefined()
            .array()
            .containsInstancesOf(Object);

        this.annotations = annotations;
        return this;
    }
}

module.exports = CsdlProperty;
