// @sap/odata-v4/lib/csdl/CsdlReturnType.js
'use strict';

const validateThat = require('../validator/ParameterValidator').validateThat;
const IllegalArgumentError = require('../errors/IllegalArgumentError');

/**
 * <a href="./../ODataSpecification/odata-v4.0-errata03-os/complete/part3-csdl/odata-v4.0-errata03-os-part3-csdl-complete.html#_Toc453752588">
 *     OData CSDL # 12.3 Element edm:ReturnType
 * </a>
 */
class CsdlReturnType {

    /**
     * @param {FullQualifiedName} type - OData CSDL # 12.3.1 Attribute Type
     */
    constructor(type) {
        validateThat('type', type).truthy().instanceOf(Object);

        /**
         * OData CSDL # 12.3.1 Attribute Type
         * @type {FullQualifiedName}
         */
        this.type = type;

        /**
         * Indicates, whether the return value is a collection
         * @type {boolean}
         */
        this.isCollection = false;

        /**
         * OData CSDL # 12.3.2 Attribute Nullable
         * @type {boolean}
         */
        this.isNullable = true;

        /**
         * OData CSDL # 6.2.2 Attribute MaxLength
         * @type {number | string}
         */
        this.maxLength = null;

        /**
         * OData CSDL # 6.2.3 Attribute Precision
         * @type {number}
         */
        this.precision = null;

        /**
         * OData CSDL # 6.2.4 Attribute Scale
         * @type {number | string}
         */
        this.scale = null;

        /**
         * OData CSDL # 14.3 Element edm:Annotation
         * @type {CsdlAnnotation[]}
         */
        this.annotations = [];
    }

    /**
     * Sets whether the return value represents a collection
     *
     * @param {boolean} isCollection - indicates whether the return value should represent a collection
     * @returns {CsdlReturnType} this instance
     */
    setCollection(isCollection) {
        validateThat('isCollection', isCollection).notNullNorUndefined().typeOf('boolean');

        this.isCollection = isCollection;
        return this;
    }

    /**
     * Sets whether the return value is nullable.
     * OData CSDL # 12.3.2 Attribute Nullable
     *
     * @param {boolean} isNullable - indicates whether the return value is nullable
     * @returns {CsdlReturnType} this instance
     */
    setNullable(isNullable) {
        validateThat('isNullable', isNullable).notNullNorUndefined().typeOf('boolean');

        this.isNullable = isNullable;
        return this;
    }

    /**
     * Sets 'MaxLength' facet.
     * OData CSDL # 6.2.2 Attribute MaxLength
     *
     * @param {number | string} maxLength - value of the 'MaxLength' facet. It must be either a
     * positive integer value or a special 'max' string value. More information can be found in the
     * CSDL specification.
     * @returns {CsdlReturnType} this instance
     */
    setMaxLength(maxLength) {
        const validate = validateThat('maxLength', maxLength).notNullNorUndefined();

        // check if the special 'max' value is specified
        if (typeof maxLength === 'string') {
            // check whether a string value is specified but not the allowed one
            if (maxLength !== 'max') {
                throw IllegalArgumentError.createForIllegalTypeValue('maxLength', 'string', 'max');
            }

            this.maxLength = 'max';
            return this;
        }

        // otherwise the value must be a positive integer
        validate.integer().positiveInteger();

        this.maxLength = maxLength;
        return this;
    }

    /**
     * Sets Precision facet.
     * OData CSDL # 6.2.3 Attribute Precision
     *
     * @param {number} precision - value of the Precision facet.
     * @returns {CsdlReturnType} this instance
     */
    setPrecision(precision) {
        validateThat('precision', precision)
            .notNullNorUndefined()
            .integer()
            .nonNegativeInteger();

        this.precision = precision;
        return this;
    }

    /**
     * Sets 'Scale' facet.
     * OData CSDL # 6.2.4 Attribute Scale
     *
     * @param {number | string} scale - value of the 'Scale' facet. It must be either a
     * non-negative integer value or a special 'variable' string value. More information can be
     * found in the CSDL specification.
     * @returns {CsdlReturnType} this instance
     */
    setScale(scale) {
        const validate = validateThat('scale', scale).notNullNorUndefined();

        // check if the special 'variable' value is specified
        if (typeof scale === 'string') {
            // check whether a string value is specified but not the allowed one
            if (scale !== 'variable') {
                throw IllegalArgumentError.createForIllegalTypeValue('scale', 'string', 'variable');
            }

            this.scale = 'variable';
            return this;
        }

        // otherwise the value must be a non-negative integer
        validate.integer().nonNegativeInteger();

        this.scale = scale;
        return this;
    }

    /**
     * OData CSDL # 6.2.6 Attribute SRID.
     * The property always has an undefined value. Modification of the property value is not
     * supported.
     * @type {number | string}
     */
    get srid() {
        return undefined;
    }

    /**
     * Sets a list of annotations for the return value.
     * OData CSDL # 14.3 Element edm:Annotation
     *
     * @param {CsdlAnnotation[]} annotations - annotations for the return value
     * @returns {CsdlReturnType} this instance
     */
    setAnnotations(annotations) {
        validateThat('annotations', annotations)
            .notNullNorUndefined()
            .array()
            .containsInstancesOf(Object);

        this.annotations = annotations;
        return this;
    }
}

module.exports = CsdlReturnType;
