// @sap/odata-v4/lib/logging/DebugLogger.js
'use strict';

const util = require('util');
const LoggerFacade = require('./LoggerFacade');

/**
 * The debug logger stores all log strings for retrieval from the debug-output serializer.
 * The originally registered logger is called additionally so nothing is lost.
 * This class has to extend LoggerFacade for context.setLogger(...) to do the right things.
 */
class DebugLogger extends LoggerFacade {
    /**
     * Creates an instance of DebugLogger.
     * @param {LoggerFacade} loggerFacade the original logger facade
     */
    constructor(loggerFacade) {
        super(loggerFacade, loggerFacade.getId());
        this._logStrings = [];
    }

    /**
     * Logging for log level debug
     * @param {Array} params The provided parameters
     * @override
     */
    debug(...params) {
        this._logger.debug(...params);
        this._logStrings.push(this._format(LoggerFacade.LogLevel.DEBUG, ...params));
    }

    /**
     * Logging for log level path
     * @param {Array} params The provided parameters
     * @override
     */
    path(...params) {
        this._logger.path(...params);
        this._logStrings.push(this._format(LoggerFacade.LogLevel.PATH, ...params));
    }

    /**
     * Logging for log level info
     * @param {Array} params The provided parameters
     * @override
     */
    info(...params) {
        this._logger.info(...params);
        this._logStrings.push(this._format(LoggerFacade.LogLevel.INFO, ...params));
    }

    /**
     * Logging for log level warning
     * @param {Array} params The provided parameters
     * @override
     */
    warning(...params) {
        this._logger.warning(...params);
        this._logStrings.push(this._format(LoggerFacade.LogLevel.WARNING, ...params));
    }

    /**
     * Logging for log level error
     * @param {Array} params The provided parameters
     * @override
     */
    error(...params) {
        this._logger.error(...params);
        this._logStrings.push(this._format(LoggerFacade.LogLevel.ERROR, ...params));
    }

    /**
     * Logging for log level fatal
     * @param {Array} params The provided parameters
     * @override
     */
    fatal(...params) {
        this._logger.fatal(...params);
        this._logStrings.push(this._format(LoggerFacade.LogLevel.FATAL, ...params));
    }

    /**
     * Format the log.
     * @param {LoggerFacade.LogLevel} level
     * @param {Array} params The provided parameters
     * @returns {string}
     * @private
     */
    _format(level, ...params) {
        const inspectOptions = { breakLength: null, depth: 4, maxArrayLength: null };
        return level + ': '
            + params.map(elem => typeof elem === 'string' ? elem : util.inspect(elem, inspectOptions)).join(' ');
    }

    /**
     * Get the collected log.
     * @returns {string[]}
     */
    getLog() {
        return this._logStrings;
    }
}

module.exports = DebugLogger;
