// @sap/odata-v4/lib/logging/LoggerFacade.js
'use strict';

const util = require('util');

const defaultLogIdGenerator = () => String(Math.ceil((Math.random() + 1) * 1000000));

/**
 * The default logging formatter. This formatter function creates a string from each log entry.
 * The string will look like:
 * [2017-07-14T12:05:00.012Z | DEBUG | 1559409]: Any debug text
 * @package
 */
const defaultFormatter = entry => {
    const inspectOptions = { breakLength: null, depth: 4, maxArrayLength: 80 };
    const message = entry.params.map(elem => typeof elem === 'string' ? elem : util.inspect(elem, inspectOptions))
        .join(' ');
    return `[${new Date(entry.timestamp).toISOString()} | ${entry.level} | ${entry.id}]: ${message}`;
};

/**
 * LogEntry definition
 * An object with this structure is created for each logging call to the LoggerFacade (e.g. info(),debug(),
 * error()) and holds data which can be logged by an logger.
 *
 * @typedef {Object} LogEntry
 * @param {number} timestamp Timestamp when the logEntry has been created (usually Date.now())
 * @param {LoggerFacade.LogLevel} level The log level
 * @param {Array} params Additional parameters to be logged
 * @param {string} id The id of the logger instance that created this log entry
 */

/**
 * The LoggerFacade wraps a provided logger instance and provides new features like formatting a
 * log entry and logging IDs.
 * @hideconstructor
 */
class LoggerFacade {
    /**
     * Creates an instance of LoggerFacade.
     * @param {Object} logger a logger instance
     * @param {number|string} [id=defaultLogIdGenerator()] Any logger id
     */
    constructor(logger, id = defaultLogIdGenerator()) {
        this._logger = logger;
        this._id = id;
        this._formatter = defaultFormatter;
    }

    /**
     * Used to output the internal status of a program.
     *
     * @param {any} params The current provided params
     * @returns {LoggerFacade} This instance
     */
    debug(...params) {
        if (this._logger) {
            const entry = this._createEntry(LoggerFacade.LogLevel.DEBUG, this._id, ...params);
            this._logger.debug(this._formatter(entry));
        }
        return this;
    }

    /**
     * Used to analyze the execution flow of a program.
     *
     * @param {any} params The current provided params
     * @returns {LoggerFacade} This instance
     */
    path(...params) {
        if (this._logger) {
            const entry = this._createEntry(LoggerFacade.LogLevel.PATH, this._id, ...params);
            this._logger.path(this._formatter(entry));
        }
        return this;
    }

    /**
     * Used for events that do not need any follow up activity.
     * This shows the normal operations within an application.
     *
     * @param {any} params The current provided params
     * @returns {LoggerFacade} This instance
     */
    info(...params) {
        if (this._logger) {
            const entry = this._createEntry(LoggerFacade.LogLevel.INFO, this._id, ...params);
            this._logger.info(this._formatter(entry));
        }
        return this;
    }

    /**
     * Used for events that need follow up activity in order to prevent errors in the future.
     *
     * @param {any} params The current provided params
     * @returns {LoggerFacade} This instance
     */
    warning(...params) {
        if (this._logger) {
            const entry = this._createEntry(LoggerFacade.LogLevel.WARNING, this._id, ...params);
            this._logger.warning(this._formatter(entry));
        }
        return this;
    }

    /**
     * Used when the desired tasks cannot be completed and the application is still usable.
     *
     * @param {any} params The current provided params
     * @returns {LoggerFacade} This instance
     */
    error(...params) {
        if (this._logger) {
            const entry = this._createEntry(LoggerFacade.LogLevel.ERROR, this._id, ...params);
            this._logger.error(this._formatter(entry));
        }
        return this;
    }

    /**
     * Used in case of errors, because of which the application is no longer usable.
     *
     * @param {any} params The current provided params
     * @returns {LoggerFacade} This instance
     */
    fatal(...params) {
        if (this._logger) {
            const entry = this._createEntry(LoggerFacade.LogLevel.FATAL, this._id, ...params);
            this._logger.fatal(this._formatter(entry));
        }
        return this;
    }

    /**
     * Set the logger.
     * @param {Object} logger a logger instance
     * @returns {LoggerFacade} This instance
     */
    setLogger(logger) {
        this._logger = logger;
        return this;
    }

    /**
     * Set the formatter function used to format each log entry. If not set a default formatting
     * function will be used which formats a log entry into a string.
     *
     * @param {Function} [formatFn=defaultFormatter] The formatter function to set
     * @returns {LoggerFacade} This instance
     */
    setFormatter(formatFn = defaultFormatter) {
        this._formatter = formatFn;
        return this;
    }

    /**
     * Returns the current logging id
     *
     * @returns {string} The current logging id
     */
    getId() {
        return this._id;
    }

    /**
     * Creates a default log entry structure used for this logger.
     *
     * @param {LoggerFacade.LogLevel} level The log level
     * @param {string} id The current logger instance id
     * @param {Array} params The provided parameters
     * @returns {LogEntry} An object structure representing the log entry
     * @private
     */
    _createEntry(level, id, ...params) {
        return {
            timestamp: Date.now(),
            level,
            params: params,
            id
        };
    }
}

/**
 * Log levels supported by the odata library
 *
 * @enum {string}
 * @readonly
 */
LoggerFacade.LogLevel = {
    DEBUG: 'DEBUG',
    PATH: 'PATH',
    INFO: 'INFO',
    WARNING: 'WARNING',
    ERROR: 'ERROR',
    FATAL: 'FATAL'
};

module.exports = LoggerFacade;
