// @sap/cds-sql/lib/sql-builder/ExpressionBuilder.js
const BaseBuilder = require('./BaseBuilder')
const {InvalidCqnObjectError} = require('../').errors

/**
 * ExpressionBuilder is used to take a part of a CQN object as an input and to build an object representing an expression
 * with SQL string and values to be used with a prepared statement.
 * The outer property 'xpr' can be omitted.
 * @example <caption>Example of xpr part of CQN </caption>
 * {
 *  xpr: [{ref: ['x']}, '<', {val: 9}]
 * }
 *
 * Each operand of the xpr can be a nested xpr.
 */
class ExpressionBuilder extends BaseBuilder {
  /**
   * The constructor of the ExpressionBuilder.
   * If the options parameter is not specified, " are used as delimiter and ? as placeholders.
   *
   * @param {object} obj - Part of the CQN object that represents an expression
   * @param {object} [options] - The configuration object.
   * @param {string} [options.delimiter] - The delimiter string.
   * @param {string} [options.placeholder] - The placeholder for prepared statement.
   * @param {string} [options.objectKey] - The object key for the expression. It can be either "xpr" or empty string.
   * Default is an empty string.
   */
  constructor (obj, options) {
    super(obj, options)
    this._options = Object.assign({objectKey: ''}, this._options)
  }

  get SelectBuilder () {
    const SelectBuilder = require('./SelectBuilder')
    Object.defineProperty(this, 'SelectBuilder', {value: SelectBuilder})
    return SelectBuilder
  }

  get ReferenceBuilder () {
    const ReferenceBuilder = require('./ReferenceBuilder')
    Object.defineProperty(this, 'ReferenceBuilder', {value: ReferenceBuilder})
    return ReferenceBuilder
  }

  /**
   * Builds an Object based on the properties of the input object in the constructor.
   * @example <caption>Example output</caption>
   * {
   *    sql: '"X" < ?',
   *    values: [1]
   * }
   *
   * @throws InvalidCqnObjectError
   * @returns {{sql: string, values: Array}} Object with two properties.
   * SQL string for prepared statement and array of values to replace the placeholders.
   */
  build () {
    this._outputObj = {
      sql: [],
      values: []
    }

    this._expressionObjectsToSQL(this._options.objectKey ? this._obj[this._options.objectKey] : this._obj)

    this._outputObj.sql = this._outputObj.sql.join(' ')
    return this._outputObj
  }

  _expressionObjectsToSQL (objects) {
    const length = objects.length
    let i = 0

    while (i < length) {
      // Some keywords need to be process as a block, while others can be treated one at a time
      if (this._reseverdKeyWords(objects, i)) {
        i = i + 3
      } else {
        this._expressionElementToSQL(objects[i])
        i++
      }
    }
  }

  /**
   * Some values and operators need to be treated in a non standard way.
   * Those are:
   * (NOT) NULL
   * (NOT) IN
   * @param {Array} objects
   * @param {number} i
   * @returns {boolean}
   * @private
   */
  _reseverdKeyWords (objects, i) {
    if (objects[i + 2] && objects[i + 2].val === null) {
      return this._addNullOrNotNull(objects[i], objects[i + 1])
    }

    if (/^(not )?in+/i.test(objects[i + 1])) {
      return this._addInOrNotIn(objects[i], objects[i + 1].toUpperCase(), objects[i + 2])
    }

    return false
  }

  /**
   * In case the value is null, SQL with reserved keywords and without a placeholder is required.
   * @param reference
   * @param operator
   * @returns {boolean}
   * @private
   */
  _addNullOrNotNull (reference, operator) {
    this._addToOutputObj(new this.ReferenceBuilder(reference, this._options).build(), false)

    this._outputObj.sql.push('IS', `${operator !== '=' ? 'NOT ' : ''}NULL`)

    return true
  }

  /**
   * (NOT) IN can have an Array or sub select instance as value.
   * @param reference
   * @param operator
   * @param values
   * @returns {boolean}
   * @private
   */
  _addInOrNotIn (reference, operator, values) {
    if (values.val instanceof Array) {
      this._addArrayForInQuery(reference, operator, values.val)
      return true
    }

    this._addSubQueryForInQuery(reference, operator, values)
    return true
  }

  _addArrayForInQuery (reference, operator, values) {
    this._addToOutputObj(new this.ReferenceBuilder(reference, this._options).build(), false)

    const placeholders = []

    for (let i = 0, length = values.length; i < length; i++) {
      placeholders.push(this._options.placeholder)
    }

    this._outputObj.sql.push(operator, '(', `${placeholders.join(', ')}`, ')')

    this._outputObj.values = this._outputObj.values.concat(values)
  }

  _addSubQueryForInQuery (reference, operator, subQuery) {
    if (this._options.objectKey) {
      delete this._options.objectKey
    }

    this._addToOutputObj(new this.ReferenceBuilder(reference, this._options).build(), false)

    this._outputObj.sql.push(operator)

    this._addToOutputObj(new this.SelectBuilder(subQuery, this._options).build(), true)
  }

  /**
   * Handles one expression element at a time.
   * @param {string|Object} element
   * @private
   */
  _expressionElementToSQL (element) {
    if (typeof element === 'string') {
      this._outputObj.sql.push(element.toUpperCase())
      return
    }

    const key = Object.keys(element)[0]

    switch (key) {
      case 'xpr':
        return this._xprOutputFromElement(element)
      case 'ref':
        return this._refOutputFromElement(element)
      case 'val':
        return this._valOutputFromElement(element)
      default:
        throw new InvalidCqnObjectError()
    }
  }

  _xprOutputFromElement (element) {
    this._options.objectKey = 'xpr'

    this._addToOutputObj(new ExpressionBuilder(element, this._options).build(), true)
  }

  /**
   * Columns come as an Object with a ref property.
   * @param {Object} element
   * @private
   */
  _refOutputFromElement (element) {
    this._addToOutputObj(new this.ReferenceBuilder(element, this._options).build(), false)
  }

  /**
   * Instead of adding the value to the SQL via string literal or string concat, add a placeholder instead.
   * The placeholder is than used by a db driver and prepared statements to defend against injections.
   * @param {Object} element
   * @private
   */
  _valOutputFromElement (element) {
    this._outputObj.sql.push(this._options.placeholder)
    this._outputObj.values.push(element.val)
  }

  _addToOutputObj ({sql, values}, addBrackets) {
    this._outputObj.sql.push(addBrackets ? `( ${sql} )` : sql)
    this._outputObj.values = this._outputObj.values.concat(values)
  }
}

module.exports = ExpressionBuilder
