// @sap/cds-services/lib/adapter/odata-v4/OData.js
/**
 * Facade for creating an instance of a EDM based OData service.
 * @alias module:odata.OData
 */
class OData {
  /**
   * Constructs an OData service for the given EDM model.
   * @param {Object} edm - the EDM model.
   * @param {Object} [options] - optional object with options.
   * @param {Object} [options.logger] - optional logger object to be used in the odata library.
   * @param {string} [options.logLevel] - optional log level to be used according to winston/npm specification.
   * @param {number} [options.maxPageSize] - The page limit value. By default server side paging is enabled and set to 100. To disable it set it to false.
   * @param {number} [options.maxResourcePathLength] - Max allowed number of resource path segments, default is 10.
   * @param {string} [options.service] - Service name as specified in CSN.
   * @param {boolean} [options.crashOnError] - Application should crash on error. Defaults to true.
   *
   * @throws ModelNotDefined - in case no or an invalid csn model is provided.
   */
  constructor (edm, options = {}) {
    this._validateEdm(edm)
    this._setOptions(options)
    this._createOdataService(edm)
  }

  _validateEdm (edm) {
    if (typeof edm !== 'object' || !edm.$Version) {
      const {ModelNotDefined} = require('../../errors')
      throw new ModelNotDefined()
    }
  }

  _setOptions (options) {
    options.maxResourcePathLength = options.maxResourcePathLength || 10
    options.debug = options.debug || false
    options.crashOnError = (options.crashOnError) === undefined ? true : Boolean(options.crashOnError)

    this._options = options
  }

  _createOdataService (edm) {
    const ServiceFactory = require('@sap/odata-v4').ServiceFactory

    this._odataService = ServiceFactory.createService(edm, require('./utils/oDataConfiguration')(edm, this._options))
    this._odataService.log(require('./utils/logger')(this._options))

    // will be added to express app like app.use('/base/path/', service) and odata-v4 wants app.use('/', service) if basePath is set
    this._odataService.setBasePath('/')
  }

  /**
   * The added cds service will be used at the handlers.
   * Some channel events have a 1:N relation to service handler events.
   * @param {Service} cdsService
   */
  addCDSServiceToChannel (cdsService) {
    this._cdsService = cdsService
    const {
      BatchExitHandler: {ATOMICITY_GROUP_END, BATCH_END},
      Components: {DATA_CREATE_HANDLER, DATA_DELETE_HANDLER, DATA_READ_HANDLER, DATA_UPDATE_HANDLER, ACTION_EXECUTE_HANDLER, LOCALE_NEGOTIATOR, METADATA_HANDLER}
    } = require('@sap/odata-v4')
    const request = require('./handlers/request')
    const {end} = require('./handlers/batch')

    this._odataService.on('error', require('./handlers/error')(this._options.crashOnError))

    if (this._options.debug) {
      this._odataService.on('debug', require('./handlers/debug'))
    }

    this._odataService.on('request', request(this._options.uaa))
    this._odataService.on(ATOMICITY_GROUP_END, end)
    this._odataService.on(BATCH_END, end)

    this._odataService.use(LOCALE_NEGOTIATOR, require('./handlers/language')(this._options))
    this._odataService.use(METADATA_HANDLER, require('./handlers/metadata')(cdsService, this._options))

    this._odataService.use(DATA_CREATE_HANDLER, require('./handlers/create')(cdsService, this._options))
    this._odataService.use(DATA_UPDATE_HANDLER, require('./handlers/update')(cdsService, this._options))
    this._odataService.use(DATA_DELETE_HANDLER, require('./handlers/delete')(cdsService, this._options))
    this._odataService.use(DATA_READ_HANDLER, require('./handlers/read')(cdsService, this._options))
    this._odataService.use(ACTION_EXECUTE_HANDLER, require('./handlers/action')(cdsService, this._options))
  }

  /**
   * Return service middleware, which can be used by node server, express, connect, ...
   * @returns {function}
   */
  getService () {
    return (req, res) => {
      const odata = () => {
        this._odataService.process(req, res)
          .catch((err) => {
            res.status(500)
            res.send({
              error: {
                code: null,
                message: err.message
              }
            })
          })
      }

      this._cdsService.routerHandle(req, res, odata)
    }
  }
}

module.exports = OData
