// @sap/odata-v4/lib/uri/apply/Transformation.js
'use strict';

/**
 * Transformation.
 * @abstract
 * @hideconstructor
 */
class Transformation {

    /**
     * Create an instance of Transformation.
     * This constructor is not intended to be called directly.  It is used in derived classes.
     * @param {Transformation.TransformationKind} kind kind of transformation
     * @protected
     */
    constructor(kind) {
        /**
         * @type {Transformation.TransformationKind}
         * @private
         */
        this._kind = kind;
    }

    /**
     * Return the transformation kind.
     * @returns {Transformation.TransformationKind} the transformation kind
     */
    getKind() {
        return this._kind;
    }
}

/**
 * Defined kinds of transformations.
 * @enum {number}
 * @readonly
 */
Transformation.TransformationKind = {
    AGGREGATE: 0,
    BOTTOM_TOP: 1,
    COMPUTE: 2,
    CONCAT: 3,
    CUSTOM_FUNCTION: 4,
    EXPAND: 5,
    FILTER: 6,
    GROUP_BY: 7,
    IDENTITY: 8,
    ORDER_BY: 9,
    SEARCH: 10,
    SKIP: 11,
    TOP: 12
};

module.exports = Transformation;
