// @sap/cds/lib/index.js
const core = require ('@sap/cds-reflect/lib/core'); (()=> require = core.lazy(require))()  // eslint-disable-line
const cds = module.exports = Object.assign ( core, {

  // Models
  resolve: require ('./models/resolve'),
  get: require ('./models/get'),
  load: lazy => cds.get.all,
  parse: require ('./models/parse'),
  unfold: require ('./models/unfold'),
  compile: require ('./models/compile'),
  localize: require ('./models/i18n'),

  // Services
  service: require ('./service/provider'),
  serve: require ('./service/serve'),
  client: require ('./service/client'),

  // Data Access
  Session: require ('./connect/session'),
  connect: require ('./connect'),
  ql: require('./connect/ql'),

  // Helpers
  in: cwd => !cwd ? cds : {__proto__:cds, cwd, config:cds.config.in(cwd) },
  cwd: lazy => process.cwd(),
  config: require('./utils/config'),
  write: require ('./utils/write'),
  require: require ('./utils/require-local'),
  version: lazy => require._('../package.json').version,
  home: lazy => require._('path').dirname (__dirname),
  build: require ('../bin/build/all')

})
cds.lazify()
