// @sap/odata-v4/lib/http/HeaderInfo.js
'use strict';

/**
 * Used to store the parts of an content type.
 * @hideconstructor
 * @ignore
 */
class HeaderInfo {

    /**
     * @param {string} name - Name
     * @param {string} value - Header value
     * @param {Buffer} [raw] - Raw header before split at =
     */
    constructor(name, value, raw) {
        this._name = name;
        this._value = value;
        this._raw = raw;
    }

    /**
     * Returns the header name
     * @returns {string} Header name
     */
    getName() {
        return this._name;
    }

    /**
     * Set the header name
     * @param {string} name - Header name
     * @returns {HeaderInfo} Header information
     * @package
     */
    setName(name) {
        this._name = name;
        return this;
    }

    /**
     * Returns the header value
     * @returns {string} header value
     */
    getValue() {
        return this._value;
    }

    /**
     * Set the header value
     * @param {string} value - Header value
     * @returns {HeaderInfo} Header information
     * @package
     */
    setValue(value) {
        this._value = value;
        return this;
    }

    /**
     * Returns the raw header
     * @returns {Buffer} raw header
     */
    getRaw() {
        return this._raw;
    }

    /**
     * Sets the raw header
     *
     * @param {Buffer} raw - raw header
     * @returns {HeaderInfo} Header information
     * @package
     */
    setRaw(raw) {
        this._raw = raw;
        return this;
    }
}

module.exports = HeaderInfo;
