// @sap/odata-v4/lib/utils/PerformanceMonitor.js
'use strict';

/**
 * This class represents a performance Monitor
 */
class PerformanceMonitor {

    /**
     * Creates an instance of PerformanceMonitor.
     *
     * @param {string} [name='default'] The name of the monitor
     */
    constructor(name = 'default') {
        this._name = name;
        this._monitors = new Map();
        this._startCpuUsage = 0;
        this._stopCpuUsage = 0;
    }

    /**
     * Returns true if this monitor is an instance of a No operation monitor.
     *
     * @returns {boolean} True if this is an instance of a no operation monitor, else false
     */
    isNoOpMonitor() {
        return false;
    }

    /**
     * Creates a child monitor.
     *
     * @param {string} [name='default'] The name of the child monitor
     * @returns {PerformanceMonitor} The child monitor
     */
    createChild(name = 'default') {
        const monitor = new PerformanceMonitor(name, true);
        this._monitors.set(name, monitor);
        return monitor;
    }

    /**
     * Returns a child monitor.
     *
     * @param {string} path A path to the requested child like "a/b/c"
     * @returns {?PerformanceMonitor} The child monitor or null if none is found
     */
    getChild(path) {
        let currentMonitor = this;
        for (const segment of path.split('/')) {
            currentMonitor = currentMonitor ? currentMonitor._monitors.get(segment) : null;
        }
        return currentMonitor;
    }

    /**
     * Returns all existing childs
     *
     * @returns {PerformanceMonitor[]} Returns all existing childs
     */
    getChilds() {
        return Array.from(this._monitors.values());
    }

    /**
     * Set the start reference time of this monitor
     *
     * @returns {PerformanceMonitor} Returns this instance
     */
    start() {
        if (!this._start) {
            this._start = process.hrtime();
            this._startCpuUsage = process.cpuUsage();
            this._startMemoryUsage = process.memoryUsage();
        }
        return this;
    }

    /**
     * Set the stop reference time of this monitor
     *
     * @returns {PerformanceMonitor} Returns this instance
     */
    stop() {
        if (!this._stop && this._start) {
            this._stop = process.hrtime(this._start);
            this._stopCpuUsage = process.cpuUsage(this._startCpuUsage);
            this._stopMemoryUsage = process.memoryUsage();
        }
        return this;
    }

    /**
     * Returns the milliseconds difference stop - start
     *
     * @returns {?number} Returns the milliseconds diiference between start and stop or null,
     *                    if monitor was not started and stopped
     */
    getMilliseconds() {
        if (this._start && this._stop) {
            // return milliseconds
            return this._stop[0] * 1e3 + this._stop[1] * 1e-6;
        }
        return 0;
    }

    /**
     * Returns the name of the monitor.
     *
     * @returns {string} The name of the monitor
     */
    getName() {
        return this._name;
    }

    /**
     * Returns the result of the monitor.
     *
     * @returns {Object} The result of the monitor
     */
    getResult() {
        const result = {};

        if (this._start && this._stop) {
            result.$Milliseconds = this.getMilliseconds();
            result['Memory usage on start (MB)'] = {
                heapTotal: this._startMemoryUsage.heapTotal / 1024 / 1024,
                heapUsed: this._startMemoryUsage.heapUsed / 1024 / 1024
            };
            result['Memory usage on stop (MB)'] = {
                heapTotal: this._stopMemoryUsage.heapTotal / 1024 / 1024,
                heapUsed: this._stopMemoryUsage.heapUsed / 1024 / 1024
            };
            result['CPU usage in milliseconds'] = {
                user: this._stopCpuUsage.user / 1000,
                system: this._stopCpuUsage.system / 1000
            };
        }

        Array.from(this._monitors.values()).forEach(
            mon => { result[mon.getName()] = mon.getResult(); });

        return result;
    }
}

/**
 * @extends PerformanceMonitor
 */
class NoOperationPerformanceMonitor extends PerformanceMonitor {

    /**
     * Creates an instance of NoOperationPerformanceMonitor.
     * @param {string} name The name of the monitor
     */
    constructor(name) {
        super(name);
    }

    /**
     * @override
     */
    isNoOpMonitor() {
        return true;
    }

    /**
     * @override
     */
    createChild() {
        return this;
    }

    /**
     * @override
     */
    getChild() {
        return this;
    }

    /**
     * @override
     */
    getChilds() {
        return [];
    }

    /**
     * @override
     */
    start() {
        return this;
    }

    /**
     * @override
     */
    stop() {
        return this;
    }

    /**
     * @override
     */
    getMilliseconds() {
        return null;
    }

    /**
     * @override
     */
    getResult() {
        return {
            name: this.getName(),
            milliseconds: 0,
            'Memory usage on start (MB)': { heapTotal: 0, heapUsed: 0 },
            'Memory usage on stop (MB)': { heapTotal: 0, heapUsed: 0 },
            'CPU usage in milliseconds': { user: 0, system: 0 }
        };
    }
}

/**
 * Creates an instance of a performance monitor.
 */
class PerformanceMonitorFactory {

    /**
     * Creates an instance of a performance monitor. If the parameter `isActive` is not true,
     * a `NoOperationPerformanceMonitor` is returned. This monitor has the same API as an active monitor,
     * but does no operation on any method call.
     *
     * @param {boolean} isActive If true an active monitor is returned, else a No operation monitor is returned
     * @param {string} [name='main'] The name of the root monitor
     * @returns {PerformanceMonitor} An instance of a PerformanceMonitor
     */
    static getInstance(isActive, name = 'main') {
        return isActive ?
            new PerformanceMonitor(name) :
            new NoOperationPerformanceMonitor(name || 'NoOperationPerformanceMonitor');
    }
}

module.exports = { PerformanceMonitorFactory };
