// @sap/cds-compiler/lib/json/nullprotos.js
let W = require("./walker");

// Adapt augmented or compacted CSN 'model' in place, by turning those objects that may
// contain properties with user-defined names (e.g. 'elements') into dictionaries, i.e.
// by setting their prototype to 'null' and return the modified model.

function nullProtos(model, options={}) {

  function nullProto(O, callback) {
    if(O !== undefined && O !== null) {
      Object.setPrototypeOf(O, null);
      if(callback)
        return [O, callback];
    }
    return undefined;
  }

  function cbEnum(/*O*/) {
    return [
    ]
  }

  function cbElement(O) {
    return [
      nullProto(O.elements, cbElement),
      nullProto(O.foreignKeys),
      nullProto(O.enum, cbEnum)
    ].concat(directItems(O.items))
  }

  function directItems(items) {
    if(items!==undefined) {
      return [
        nullProto(items.elements, cbElement),
        nullProto(items.enum, cbEnum)
      ]
    }
    return undefined;
  }

  function cbArtifact(O) {
    return [
      nullProto(O.artifacts, cbArtifact),
      nullProto(O.elements, cbElement),
      nullProto(O.actions, cbAction),
    ]
  }

  function cbParam(/*O*/) {
    return [
    ]
  }

  function cbAction(O) {
    return [
      nullProto(O.params, cbParam),
      nullProto(O.returns && O.returns.elements, cbElement),
      nullProto(O.returns && O.returns.enum, cbEnum)
    ]
  }

  function cbDefinition(O) {
    return [
      nullProto(O.elements, cbElement),
      nullProto(O.enum, cbEnum),
      nullProto(O.artifacts, cbArtifact),
      nullProto(O.actions, cbAction),
      nullProto(O.params, cbParam),
      nullProto(O.foreignKeys),
      nullProto(O.returns && O.returns.elements, cbElement),
      nullProto(O.returns && O.returns.enum, cbEnum)
    ].concat(directItems(O.items))
  }

  function doSetAllMissingProtos(o) {
    W.walk(o, (isNode, N) => {
      if(isNode && Object.getPrototypeOf(N)===null && !Array.isArray(N))
        Object.setPrototypeOf(N,Object.prototype)
    })
  }

  if(options.setAllMissingProtos)
    doSetAllMissingProtos(model);

  nullProto(model.definitions),
  W.walkNodesExFn(model.definitions, cbDefinition);
  return model;
}

module.exports = nullProtos;
