// @sap/cds/bin/dev.js
module.exports = Object.assign ( dev, {
  flags: [
    '--dry'
  ],
  help: `
# SYNOPSIS

    *cds dev* <command> [<dest>]

    Setup and manage an umbrella project for developing *cds* with all
    sub projects cloned locally using an umbrella-level node_modules lib
    and cleaning up sub projects to use _standard semver_ dependencies,
    _no shrinkwrap_ files and _no .npmrc_ files before each \`npm install\`.


# COMMANDS

  *s* | *setup*      ... local umbrella by git-cloning sub projects
  *p* | *pull*       ... latest commits from sub projects' repos
  *b* | *backup*     ... dependencies and .npmrc files
  *c* | *cleanup*    ... dependencies to std semver
  *r* | *restore*    ... sub projects from backup.zip
  *x* | *reset*      ... by rm -rf all node_modules folders

`})


const fs = require ('../lib/utils/fs.js'), {path} = fs
const { promisify } = require('util')
const { exec } = require('child_process')
const write = promisify (fs.writeFile)
const run = promisify (exec)

const shortcuts = {
    s: 'setup',
    p: 'pull',
    b: 'backup',
    c: 'cleanup',
    r: 'restore',
    x: 'reset',
}


function dev([cmd, dest]){
if (dest)  return fs.mkdirp (dest, ()=>{
    console.log ('\n    Creating project folder:', dest)
    process.chdir(dest); dev(cmd)
})

const pck = (()=>{try{
    return require (path.resolve('package.json'))
}catch(e){
    const name = path.basename(process.cwd())
    const package_json = {
        name,  version: '0.1.0',
        author: '...', description: '...', repository: '...',
    }
    if (name === 'cds' || name === 'cdx') Object.assign (package_json, {
        devDependencies: {
            jest: '*', 'jest-junit': '*', supertest: '*', hasbin:'*',
            sqlite: 'latest', express: 'latest',
        },
        jest: {
            roots: [ 'central', 'core' ],
            collectCoverageFrom: [ 'central/lib/**', 'core/lib/**' ],
            testPathIgnorePatterns: [ '/node_modules/', '/integration/']
        },
        scripts: {
            'test:new': 'jest --testRegex test-new.js',
            test: 'jest && npm run test:new',
        },
      })
    return package_json
}})()

const done = ()=> process.stdout.write ('done.\n')
const fetch_package_jsons = ()=> run('ls */package.json')
.then (({stdout}) => stdout.split('\n').slice(0,-1))

const commands = {
    ls: ()=> fetch_package_jsons() .then (all => all.join('\n')) .then (console.log),
    setup, pull, backup, cleanup, restore, reset
}

!(
    commands [cmd]
    || commands [shortcuts[cmd]]
    || _invalid
)()


function _invalid(){
    console.error ('Invalid command: '+ cmd)
}

function setup(){

    let isnew = false
    const projects = pck['sub-projects'] || (pck['sub-projects'] =
        /cd[sx]/.test (pck.name) ? require('./.dev-internal.js') : isnew = {
        ">>>>": "specify your sub projects with local folder and git repo link as in the examples below...",
        "sub-project-one": "https://github.com/repo-one",
        "second-project": "https://github.com/repo-two",
    })
    if (isnew || projects[">>>>"])  return write (path.resolve('package.json'), JSON.stringify(pck,null,'  '))
    .then (()=> console.log(`
    Please specifiy your sub projects with local folders
    and git repo links in the created package.json.
    `))

    clone(projects) .then (()=>{
        const scripts = pck.scripts || (pck.scripts = {})
        if (!scripts.setup) scripts.setup = "cds dev setup"
        if (!scripts.preinstall)  scripts.preinstall = "cds dev cleanup"
        if (!scripts.postinstall)  scripts.postinstall = "cds dev restore"
        if (!scripts.reset) scripts.reset = "cds dev reset && npm i --prod -s"
        return write (path.resolve('package.json'), JSON.stringify(pck,null,'  '))
    }) .then (install) .then (()=>console.log('\n/all done.'))

}


function clone (projects) {

    console.log ('\nCloning sub projects and adding local dependencies... \n')
    if (!pck.dependencies)  pck.dependencies = {}
    let cloning = false

    const promise = Promise.all (Object.entries(projects).map (([dest,url]) => new Promise ((_resolved, _reject) => {
        let [repo, branch='master'] = url.replace('git://','https://').split('#')
        if (fs.isdir(dest))  return _resolved (console.log ('  ./'+dest, 'exists'));  else cloning = true
        let p = exec (`git clone -b ${branch} ${repo} ${dest}`)
        p.stdout.on ('data', d=>process.stdout.write('  '+d))
        p.stderr.on ('data', d=>process.stderr.write('  '+d))
        p.on ('close', _resolved) .on ('error', _reject)
    }).then(()=>{
        const p = require (path.resolve(dest,'package.json'))
        pck.dependencies [p.name] = 'file:'+ dest
    })))

    if (cloning)  setTimeout (()=>{
        process.stdout.write ('\ntakes about 20 sec, please be patient... ')
    }, 500)
    else console.log ('\n  ( No cloning required; use `cds dev pull` to update from git repos. )')

    return promise
}


function install(){ return new Promise ((_resolved, _reject) => {
    console.log ('\nRunning npm install... \n')
    let p = exec ('npm i --prod -s', (e) => e ? _reject(e) : _resolved())
    p.stdout.on ('data', d=>process.stdout.write (d === 'done.\n' ? d : '  '+d))
    p.stderr.on ('data', d=>process.stderr.write (d === 'done.\n' ? d : '  '+d))
})}


function pull(){
    process.stdout.write ('Pulling latest updates from sub-projects... ')
    const root = process.cwd(),  projects = pck['sub-projects']
    return Object.entries(projects) .reduce ((prev,[dest,url]) => prev.then (()=>{
        let [repo, branch='master'] = url.replace('git://','https://').split('#')
        process.chdir (path.resolve(root,dest))
        let pull = run (`git pull`)
        if (repo.match(/cds-compiler/) && branch === 'alpha')
            pull = pull.then (({stdout})=> /Already up to date/.test(stdout) || run(`git reset --hard HEAD~`))
        return pull
    }), Promise.resolve()) .then(done) .then (()=>process.chdir(root))
}

function backup (package_jsons) {
    if (!package_jsons)  return fetch_package_jsons() .then (backup)
    process.stdout.write ('Backing up sub projects to backup.zip ... ')
    return [ run ('rm -f backup.zip'),
        `find . -name npm-shrinkwrap.json -exec zip backup.zip {} \\;`,
        `find . -name .npmrc -exec zip backup.zip {} \\;`,
        `zip backup.zip `+ package_jsons.join(' '),
    ] .reduce ((prev,next) => prev.then (()=>run(next)))
    .then (done) .then (()=>package_jsons)
}

function cleanup(){
    return fetch_package_jsons() .then (backup) .then (package_jsons => {
        process.stdout.write ('Cleaning up sub projects... ')
        const deps = pck.dependencies
        for (let package_json of package_jsons) {
            const pjson = require (path.resolve(package_json))
            for (let d in pjson.dependencies) {
                if (/^file:/.test(deps[d]))  pjson.dependencies[d] = '*'
            }
            write (package_json, JSON.stringify(pjson,null,'  '))
        }
    })
    .then (()=> Promise.all ([
        `find . -name npm-shrinkwrap.json -exec rm {} \\;`,
        `find . -name .npmrc -exec rm {} \\;`,
    ] .map (each => run(each))))
    .then (done)
}

function restore(){
    process.stdout.write ('Restoring sub projects from backup... ')
    run('unzip -o backup.zip') .then (done)
}

function reset(){
    process.stdout.write ('Deleting `node_modules` in parent and sub projects... ')
    run('rm -fr node_modules */node_modules') .then (done)
}

}/* eslint no-console:0 */
