// @sap/odata-v4/client/lib/uri/builder/ODataV4UriBuilder.js
'use strict';

const ResourceKind = require('../../../../lib/uri/UriResource').ResourceKind;
const SemanticODataUriBuilder = require('./SemanticOdataUriBuilder');
const ParameterKind = require('./SegmentParameter').ParameterKind;
const QueryOptions = require('../../../../lib/uri/UriInfo').QueryOptions;

class V4SemanticODataUriBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} namespace
     * @param {string} name
     * @return {V4FunctionBuilder}
     */
    boundFunction(namespace, name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.BOUND_FUNCTION, namespace + '.' + name);
        // eslint-disable-next-line no-use-before-define
        return new V4FunctionBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} search
     * @return {V4SemanticODataUriBuilder}
     */
    search(search) {
        this._odataUriBuilder.setQueryOption(QueryOptions.SEARCH, search);
        return this;
    }
}

class V4PrimitivePropertyBuilder extends V4SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    value() {
        this._odataUriBuilder.addPathSegment(ResourceKind.VALUE);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V4CollectionPropertyBuilder extends V4SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    count() {
        this._odataUriBuilder.addPathSegment(ResourceKind.COUNT);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

}

class V4ComplexPropertyBuilder extends V4SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V4ComplexPropertyBuilder}
     */
    complexProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.COMPLEX_PROPERTY, name);
        return this;
    }

    /**
     * @param {string} name
     * @returns {V4CollectionPropertyBuilder}
     */
    complexCollectionProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.COMPLEX_COLLECTION_PROPERTY, name);
        return new V4CollectionPropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V4CollectionPropertyBuilder}
     */
    primitiveCollectionProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.PRIMITIVE_COLLECTION_PROPERTY, name);
        return new V4CollectionPropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V4PrimitivePropertyBuilder}
     */
    primitiveProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.PRIMITIVE_PROPERTY, name);
        return new V4PrimitivePropertyBuilder(this._odataUriBuilder);
    }
}

class V4EntityBuilder extends V4SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     */
    navigateCollection(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.NAVIGATION_TO_MANY, name);
        // eslint-disable-next-line no-use-before-define
        return new V4EntityCollectionBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     */
    navigateSingle(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.NAVIGATION_TO_ONE, name);
        return this;
    }

    /**
     * @param {string} name
     * @returns {V4ComplexPropertyBuilder}
     */
    complexProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.COMPLEX_PROPERTY, name);
        return new V4ComplexPropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V4CollectionPropertyBuilder}
     */
    primitiveCollectionProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.PRIMITIVE_COLLECTION_PROPERTY, name);
        return new V4CollectionPropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V4CollectionPropertyBuilder}
     */
    complexCollectionProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.COMPLEX_COLLECTION_PROPERTY, name);
        return new V4CollectionPropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V4PrimitivePropertyBuilder}
     */
    primitiveProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.PRIMITIVE_PROPERTY, name);
        return new V4PrimitivePropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    value() {
        this._odataUriBuilder.addPathSegment(ResourceKind.VALUE);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @returns {SemanticODataUriBuilder}
     */
    ref() {
        this._odataUriBuilder.addPathSegment(ResourceKind.REF);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} namespace
     * @param {string} name
     * @returns {SemanticODataUriBuilder}
     */
    boundAction(namespace, name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.BOUND_ACTION, namespace + '.' + name);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V4KeyBuilder extends V4EntityBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {string} [type]
     * @returns {V4KeyBuilder}
     */
    compositeKey(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, name, value, type);
        return this;
    }
}

class V4FunctionBuilder extends V4KeyBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V4FunctionBuilder}
     */
    parameter(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.FUNCTION_PARAMETER, name, value, type);
        return this;
    }

    /**
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V4EntityBuilder}
     */
    key(value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, null, value, type);
        return new V4EntityBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V4KeyBuilder}
     */
    compositeKey(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, name, value, type);
        return new V4KeyBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    count() {
        this._odataUriBuilder.addPathSegment(ResourceKind.COUNT, '$count');
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V4EntityCollectionBuilder extends V4SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V4EntityBuilder}
     */
    key(value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, null, value, type);
        return new V4EntityBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V4KeyBuilder}
     */
    compositeKey(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, name, value, type);
        return new V4KeyBuilder(this._odataUriBuilder);
    }

    /**
     * @returns {SemanticODataUriBuilder}
     */
    count() {
        this._odataUriBuilder.addPathSegment(ResourceKind.COUNT);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @returns {SemanticODataUriBuilder}
     */
    ref() {
        this._odataUriBuilder.addPathSegment(ResourceKind.REF_COLLECTION);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} namespace
     * @param {string} name
     * @returns {SemanticODataUriBuilder}
     */
    boundAction(namespace, name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.BOUND_ACTION, namespace + '.' + name);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class ODataV4UriBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     */
    entitySet(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.ENTITY_COLLECTION, name);
        return new V4EntityCollectionBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V4EntityBuilder}
     */
    singleton(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.SINGLETON, name);
        return new V4EntityBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V4FunctionBuilder}
     */
    functionImport(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.FUNCTION_IMPORT, name);
        return new V4FunctionBuilder(this._odataUriBuilder);
    }


    /**
     * @param {string} name
     * @returns {SemanticODataUriBuilder}
     */
    actionImport(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.ACTION_IMPORT, name);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    batch() {
        this._odataUriBuilder.addPathSegment(ResourceKind.BATCH, '$batch');
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    metadata() {
        this._odataUriBuilder.addPathSegment(ResourceKind.METADATA, '$metadata');
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

module.exports = ODataV4UriBuilder;
