// @sap/odata-v4/client/lib/uri/builder/ODataV2UriBuilder.js
'use strict';
const ResourceKind = require('../../../../lib/uri/UriResource').ResourceKind;
const SemanticODataUriBuilder = require('./SemanticOdataUriBuilder');
const ParameterKind = require('./SegmentParameter').ParameterKind;

class V2PropertyBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    value() {
        this._odataUriBuilder.addPathSegment(ResourceKind.VALUE);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V2ComplexPropertyBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @returns {V2ComplexPropertyBuilder}
     */
    complexProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.COMPLEX_PROPERTY, name);
        return this;
    }

    /**
     * @param {string} name
     * @return {V2PropertyBuilder}
     */
    primitiveProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.PRIMITIVE_PROPERTY, name);
        return new V2PropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    value() {
        this._odataUriBuilder.addPathSegment(ResourceKind.VALUE);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V2LinkBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    navigateSingle(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.NAVIGATION_TO_ONE, name);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    navigateCollection(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.NAVIGATION_TO_MANY, name);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

}

class V2EntityBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V2EntityCollectionBuilder}
     */
    navigateCollection(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.NAVIGATION_TO_MANY, name);
        // eslint-disable-next-line no-use-before-define
        return new V2EntityCollectionBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V2EntityBuilder}
     */
    navigateSingle(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.NAVIGATION_TO_ONE, name);
        return this;
    }

    /**
     * @return {V2LinkBuilder}
     */
    links() {
        this._odataUriBuilder.addPathSegment(ResourceKind.REF);
        return new V2LinkBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V2ComplexPropertyBuilder}
     */
    complexProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.COMPLEX_PROPERTY, name);
        return new V2ComplexPropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V2PropertyBuilder}
     */
    primitiveProperty(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.PRIMITIVE_PROPERTY, name);
        return new V2PropertyBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    value() {
        this._odataUriBuilder.addPathSegment(ResourceKind.VALUE);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V2KeyBuilder extends V2EntityBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {string} [type]
     * @returns {V2KeyBuilder}
     */
    compositeKey(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, name, value, type);
        return this;
    }
}

class V2EntityCollectionBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V2EntityBuilder}
     */
    key(value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, null, value, type);
        return new V2EntityBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V2KeyBuilder}
     */
    compositeKey(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, name, value, type);
        return new V2KeyBuilder(this._odataUriBuilder);
    }

    count() {
        this._odataUriBuilder.addPathSegment(ResourceKind.COUNT);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

class V2FunctionImportBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string|number|boolean} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V2FunctionImportBuilder}
     */
    parameter(name, value, type) {
        this._odataUriBuilder.addV2FunctionParameter(name, value, type);
        return this;
    }

    /**
     * @param {string} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V2EntityBuilder}
     */
    key(value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, null, value, type);
        return new V2EntityBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @param {string} value
     * @param {EdmPrimitiveType} [type]
     * @returns {V2KeyBuilder}
     */
    compositeKey(name, value, type) {
        this._odataUriBuilder.addToSegment(ParameterKind.KEY_PREDICATE, name, value, type);
        return new V2KeyBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    value() {
        this._odataUriBuilder.addPathSegment(ResourceKind.VALUE);
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    count() {
        this._odataUriBuilder.addPathSegment(ResourceKind.COUNT, '$count');
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

}

class ODataV2UriBuilder extends SemanticODataUriBuilder {

    /**
     * @param {ODataUriBuilder} odataUriBuilder
     */
    constructor(odataUriBuilder) {
        super(odataUriBuilder);
    }

    /**
     * @param {string} name
     */
    entitySet(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.ENTITY_COLLECTION, name);
        return new V2EntityCollectionBuilder(this._odataUriBuilder);
    }

    /**
     * @param {string} name
     * @return {V2FunctionImportBuilder}
     */
    functionImport(name) {
        this._odataUriBuilder.addPathSegment(ResourceKind.FUNCTION_IMPORT, name);
        return new V2FunctionImportBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    batch() {
        this._odataUriBuilder.addPathSegment(ResourceKind.BATCH, '$batch');
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }

    /**
     * @return {SemanticODataUriBuilder}
     */
    metadata() {
        this._odataUriBuilder.addPathSegment(ResourceKind.METADATA, '$metadata');
        return new SemanticODataUriBuilder(this._odataUriBuilder);
    }
}

module.exports = ODataV2UriBuilder;
