// @sap/odata-v4/client/lib/uri/builder/SegmentParameter.js
'use strict';

class SegmentParameter {

    /**
     * @param {SegmentParameter.ParameterKind} parameterKind
     * @param {*} value
     */
    constructor(parameterKind) {
        this._parameterKind = parameterKind;
        this._value = null;
        this._name = null;
        this._type = null;
    }

    /**
     * @param {EdmPrimitiveType} primitiveType
     * @return SegmentParameter
     */
    setType(primitiveType) {
        this._type = primitiveType;
        return this;
    }

    /**
     * @param {string} name
     * @return SegmentParameter
     */
    setName(name) {
        this._name = name;
        return this;
    }

    /**
     * @param {*} value
     * @return SegmentParameter
     */
    setValue(value) {
        this._value = value;
        return this;
    }

    /**
     * @returns {SegmentParameter.ParameterKind}
     */
    getParameterKind() {
        return this._parameterKind;
    }

    /**
     * @returns {*}
     */
    getValue() {
        return this._value;
    }

    /**
     * @returns {?string}
     */
    getName() {
        return this._name;
    }

    /**
     * @return {null|EdmPrimitiveType}
     */
    getType() {
        return this._type;
    }
}

/**
 * @readonly
 * @enum {string}
 */
SegmentParameter.ParameterKind = {
    FUNCTION_PARAMETER: 'function-parameter',
    KEY_PREDICATE: 'key-predicate'
};

module.exports = SegmentParameter;
