// @sap/odata-v4/lib/format/JsonFormat.js
'use strict';

/**
 * Stores information specific to the JSON format.
 */
class JsonFormat {
}
/**
 * Annotations
 *
 * @enum {string}
 * @readonly
 */
JsonFormat.Annotations = {
    CONTEXT: '@odata.context',
    COUNT: '@odata.count',
    ETAG: '@odata.etag',
    METADATA_ETAG: '@odata.metadataEtag',
    MEDIA_CONTENT_TYPE: '@odata.mediaContentType',
    BIND: '@odata.bind',
    ID: '@odata.id',
    TYPE: '@odata.type',
    NEXT_LINK: '@odata.nextLink',
    getMetaProperty(annotation) {
        switch (annotation) {
            case JsonFormat.Annotations.CONTEXT:
                return JsonFormat.MetaProperties.CONTEXT;
            case JsonFormat.Annotations.COUNT:
                return JsonFormat.MetaProperties.COUNT;
            case JsonFormat.Annotations.ETAG:
                return JsonFormat.MetaProperties.ETAG;
            case JsonFormat.Annotations.METADATA_ETAG:
                return JsonFormat.MetaProperties.METADATA_ETAG;
            case JsonFormat.Annotations.MEDIA_CONTENT_TYPE:
                return JsonFormat.MetaProperties.MEDIA_CONTENT_TYPE;
            case JsonFormat.Annotations.BIND:
                return JsonFormat.MetaProperties.BIND;
            case JsonFormat.Annotations.ID:
                return JsonFormat.MetaProperties.ID;
            case JsonFormat.Annotations.TYPE:
                return JsonFormat.MetaProperties.TYPE;
            case JsonFormat.Annotations.NEXT_LINK:
                return JsonFormat.MetaProperties.NEXT_LINK;
            default:
                return null;
        }
    }
};

/**
 * Internal meta properties
 *
 * @enum {string}
 * @readonly
 */
JsonFormat.MetaProperties = {
    CONTEXT: '*@odata.context',
    COUNT: '*@odata.count',
    ETAG: '*@odata.etag',
    METADATA_ETAG: '*@odata.metadataEtag',
    MEDIA_CONTENT_TYPE: '*@odata.mediaContentType',
    BIND: '*@odata.bind',
    ID: '*@odata.id',
    TYPE: '*@odata.type',
    NEXT_LINK: '*@odata.nextLink',
    getAnnotation(metaProperty) {
        switch (metaProperty) {
            case JsonFormat.MetaProperties.CONTEXT:
                return JsonFormat.Annotations.CONTEXT;
            case JsonFormat.MetaProperties.COUNT:
                return JsonFormat.Annotations.COUNT;
            case JsonFormat.MetaProperties.ETAG:
                return JsonFormat.Annotations.ETAG;
            case JsonFormat.MetaProperties.METADATA_ETAG:
                return JsonFormat.Annotations.METADATA_ETAG;
            case JsonFormat.MetaProperties.MEDIA_CONTENT_TYPE:
                return JsonFormat.Annotations.MEDIA_CONTENT_TYPE;
            case JsonFormat.MetaProperties.BIND:
                return JsonFormat.Annotations.BIND;
            case JsonFormat.MetaProperties.ID:
                return JsonFormat.Annotations.ID;
            case JsonFormat.MetaProperties.TYPE:
                return JsonFormat.Annotations.TYPE;
            case JsonFormat.MetaProperties.NEXT_LINK:
                return JsonFormat.Annotations.NEXT_LINK;
            default:
                return null;
        }
    }
};

module.exports = JsonFormat;
